These are scripts for strain analysis and assesment for P.malariae
Run the scripts in the following order

1. m2_pm_bcodes.Rmd - Extract the relevant barcodes for the Pm cells in mixed and pure infections.
2. soupc_v25_m2.nf - Run souporcell using 2 alternative aligners (Minimap and hisat-2). This script calls the following scripts
   - soupc_v25_mnmap_hsat_310.sh - alignment script
   - I ran the script like this pointing to relevant paths: `nextflow run soupc_v25_m2.nf --id_decode "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/raw/Pm/pm_solo_mixed_mdata_decode.csv" --bam "/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/cellranger_runs/Pm/*/outs/possorted_genome_bam.bam" --bcodes "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/raw/Pm/MSC*/outs/sk21_bcodes_pp/barcodes.tsv.gz" --o_dir "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pm/" --soup_dir "soupc" --scrpt "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/multipurpose_scripts/soupc_v25_mnmap_hsat_310.sh" --ref_file "PlasmoDB-66_PmalariaeUG01_Genome.fasta" --hsat_ref_file "Pm66" -resume`
3. m2_pm_strain_decon_all.Rmd - Assess sourpocell clustering for a range of Ks by both minimap and hisat2 using knee plots, heatmaps and sankey plots. This allows to narrow down to a smaller range of likely Ks
4. pbulk_gtyping.nf - Pseudobulk the cells into the clusters assignments for the likely Ks and genotype using freebayes and count the number of reads/molecules supporting each allele using vartrix. calls
   - cr_subset_bam_linux.sh - subseting cells from bam files and creating new pseudobulk categories
   - I ran the script like this pointing to relevant paths : `nextflow run pbulk_gtyping.nf --sv_dir "pbulk_gtypes_preQC" --bam "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pm/MSC*/soupc/*/parent/possorted_genome_bam.bam"  --bcodes "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pm/MSC*/pbulk_gtypes_preQC/bcodes/minmap/*/*tsv" --odir "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pm/" --mapper "minmap" --scrpt "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/multipurpose_scripts/cr_subset_bam_linux.sh" --ref  "/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/references_for_souporcell/PlasmoDB-66_PmalariaeUG01_Genome.fasta" -resume` 
5. m2_pm_strain_deconv_refnd_all.Rmd - Assess the most accurate K by looking at the genotypes of the pseudobulk clusters using heatmaps and chromosome plots. Also assess relatedness
6. Further things to do
   - Start with long read RNA
   - Use the new reference files generated by Sunil - removing the hypervariable regions
   - Rerun the pseudobulk genotyping with further stratification based on stages
   - Assess DE between strains.

Note
- You may need to change the "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/multipurpose_scripts/" to your_dir"bash_scripts/" to access bash scripts being called by the nextflow script.
- hisat2_ref_build.sh - needed to generate hisat appropriate references

   
