These are scripts for strain analysis and assesment for P.malariae
Run the scripts in the following order

1. m2_pm_bcodes.Rmd - Extract the relevant barcodes for the Pm cells in mixed and pure infections using the script
2. soupc_v25_m2.nf - Run souporcell using 2 alternative aligners (Minimap and hisat-2) using the script. This script calls the following scripts
   - soupc_v25_mnmap_hsat_310.sh - alignment script
3. m2_pm_strain_decon_all.Rmd - Assess sourpocell clustering for a range of Ks by both minimap and hisat2 using knee plots, heatmaps and sankey plots. This allows to narrow down to a smaller range of likely Ks
4. pbulk_gtyping.nf - Pseudobulk the cells into the clusters assignments for the likely Ks and genotype using freebayes and count the number of reads/molecules supporting each allele using vartrix. calls
   - cr_subset_bam_linux.sh - subseting cells from bam files and creating new pseudobulk categories
5. m2_pm_strain_deconv_refnd_all.Rmd - Assess the most accurate K by looking at the genotypes of the pseudobulk clusters using heatmaps and chromosome plots. Also assess relatedness
6. Further things to do
   - Start with long read RNA
   - Use the new reference files generated by Sunil - removing the hypervariable regions
   - Rerun the pseudobulk genotyping with further stratification based on stages
   - Assess DE between strains.

   
