---
title: "msc_strain_decon_new"
date: '2023-08-10'
author: 'Jesse Rop'
output:
  html_document:
    toc: true
    theme: united
    code_folding: hide
    df_print: paged
---


```{r, message=FALSE}
suppressPackageStartupMessages({
  library(Seurat)
  library(dplyr) 
  library(stringr)
  library(purrr)
  library(readr)
  library(tibble)
  library(ggplot2)
  library(tidyr)
  library(cowplot)
  library(pheatmap)
  # library(rmarkdown)
  # library('vcfR')
  library(conflicted)
  # library(RColorBrewer)
  # library(MetBrewer)
  library(ComplexHeatmap)
  # library(SeqArray)
  # library(ggrepel)
  library(ggvenn)
  # library(magick)
  # library(svglite)
  # library(plotly)
  library(ggsankey)
  # library(gganimate)
  # library(av)
  
  conflict_prefer("select", "dplyr")
  conflict_prefer("count", "dplyr")
  conflict_prefer("filter", "dplyr")
  conflict_prefer("rename", "dplyr")
  conflict_prefer("saveRDS", "base")
  conflicts_prefer(base::as.data.frame)
})

```


# MSC parasites strain deconvolution and VCF visualization 

## Variable declarations

 Common functions

### Source common functions, varriables and tables
```{r}
##Boxplot labeling function
source("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/multipurpose_scripts/seurat_sce/seu_sce_common_fns.R")
source("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/multipurpose_scripts/plotting_common_fns.R")
source("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/mali22_code/pf_common_vars.R")
```

### Declare variables

```{r}
sample_set = 'Mali2'
algn_nm = 'minmap'
species = 'Pm'

soup_dir = "soupc"
sv_dir = "pbulk_gtypes_preQC"

# soup_dir = "soupc_GE_postQC"
# sv_dir = "pbulk_gtypes_GE_postQC"

strain_levels_univrs <- c(paste0("SC", 1:20), "Doublet", "Negative")
```

```{r setup}
knitr::opts_knit$set(root.dir = '/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/')
```



```{r}
##Copy souporcell files for the optimum clusters

## Sample IDs
sample_name_nms <- c("MSC27", "MSC35", "MSC42", "MSC44", "MSC47", "MSC36", "MSC16", "MSC15") %>% sort()

## Mapping techniques
algn_nms = c('minmap', 'hsat')

## Match donors with mapping technique
sample_name_nms_al <- rep(sample_name_nms, length(algn_nms))
algn_nms_al <- rep(algn_nms, each= length(sample_name_nms))

sm_aln_nms_al <- paste(rep(sample_name_nms, length(algn_nms)), rep(algn_nms, each= length(sample_name_nms)), sep ='_')

## number of optimum clusters to investigate per sample
opt_clusters_nms <- list("MSC27" = c(1:5), "MSC35" = c(2:6), "MSC42" = c(1:5), "MSC44" = c(1:5), "MSC47" = c(4:8), "MSC36" = c(1:5), "MSC16" = c(1:5), "MSC15" = c(2:6)) %>% .[sample_name_nms]

opt_clusters_len <- map(opt_clusters_nms, length)

## number of optimum clusters to investigate per sample factoring the 2 alignments
opt_clusters_nms_aln <- rep(opt_clusters_nms, length(algn_nms)) %>% set_names(sm_aln_nms_al)
opt_clusters_len_aln <- map(opt_clusters_nms_aln, length)

##Samples
sample_name <- map2(sample_name_nms, opt_clusters_len[sample_name_nms], ~rep(.x, each = .y))

opt_clusters <- unlist(flatten(opt_clusters_nms[sample_name_nms]))

sample_name <- unlist(flatten(sample_name))

## sorted donor_opt_cluster names
sorted_names <- paste0(sample_name, "_", opt_clusters)

## sorted factoring alignment
sorted_names_al <-paste0(sample_name, "_",rep(algn_nms,each=length(sample_name)),"_", opt_clusters)

iter = opt_clusters_len

```

## Assess accurate K

### log likelihood plots

Assess range of accurate K from log likelihood plots
```{r, message=FALSE}
## Read in log likelihoods for estimation of number of clusters from elbow plots
clusters_log_likelihoods_lst = map2(sample_name_nms_al, rep(algn_nms, each= length(sample_name_nms)), ~read_table(paste0('data/processed/',species,'/', .x,'/', soup_dir, '/',.y,'/clusters_log_likelihoods.txt'), col_names = c('No_of_clusters', 'td1','td2','td3', 'td4','log_likelihood')) %>% select(-(starts_with("td"))) %>% mutate(across(No_of_clusters, ~str_remove_all(., ".*out_dir_k|/clusters.err:best")))) %>% set_names(paste(sample_name_nms_al, rep(algn_nms, each= length(sample_name_nms)), sep ='_'))
```


```{r} 
ll_plts <- map(clusters_log_likelihoods_lst, ~.x %>%
  mutate(across(c('No_of_clusters'), ~factor(as.numeric(.)))) %>%
  ggplot(., aes(x = No_of_clusters, y = log_likelihood)) +
  geom_point() +
  geom_line(aes(group = 1)) +
  # facet_grid(rows = .~id)+
  theme(axis.text=element_text(size=10),
        axis.title=element_text(size=16,face="bold"),
        legend.text = element_text(size=15),
        title = element_text(size=16,face="bold"),
        strip.text.x = element_text(size = 18, face = "bold")))
```

#### minimap log likelihood plots {.tabset} 
```{r, results='asis'}

iwalk(ll_plts[str_detect(names(ll_plts), "minmap")],~{
        cat('##### ',.y,'   \n')
        print(.x)
        cat('\n \n')
      })
```


#### hisat2 log likelihood plots {.tabset} 
```{r, results='asis'}

iwalk(ll_plts[str_detect(names(ll_plts), "hsat")],~{
        cat('##### ',.y,'   \n')
        print(.x)
        cat('\n \n')
      })
```

### log likelihood plots - same axis

```{r, fig.width= 13, fig.height=6} 
clusters_log_likelihoods_lst %>% 
  bind_rows(., .id = "sample_algnr") %>% 
  separate_wider_regex(cols = "sample_algnr", c(sample = "MSC[0-9]*|MSC[0-9]*_[A-Z]*", "_",algnr = "[a-z].*")) %>%
  # separate_wider_delim(., "sample_algnr", names = c("sample", "algnr"), delim = "_") %>% 
  mutate(across(c('sample'), ~factor(., levels = sample_name_nms))) %>%
  mutate(across(c('No_of_clusters'), ~factor(as.numeric(.)))) %>%
  ggplot(., aes(x = No_of_clusters, y = log_likelihood)) +
    geom_point() +
    geom_line(aes(group = 1)) + facet_grid(algnr ~ sample, scales = "free_y")
```



Read all the souporcell files for all the iterations. May be excessive but since we have the memory and cpu and we may need to use the files downstream then why not.
```{r, results='hide', eval=FALSE}
# list.files(paste0("data/processed/", species,"/"), pattern = "clusters.tsv", full.names = T, recursive = T, include.dirs = T)
list.files(paste0("data/processed/", species), pattern = "clusters.tsv", full.names = T, recursive = T, include.dirs = T)%>% set_names(nm = (str_remove_all(.,paste0('data/processed/', species, '/|soupc/|k|/clusters.tsv')) %>% str_replace_all(., '/', '_'))) 
```

Use when only running minimap2
```{r, eval=FALSE}
soupc_all <- list.files(paste0("data/processed/", species), pattern = "clusters.tsv", full.names = T, recursive = T, include.dirs = T)%>%
  set_names(nm = (str_remove_all(.,paste0('data/processed/', species, '/|soupc/m.*k|/k|/clusters.tsv')) %>% str_replace_all(., '/', '_'))) %>% .[sorted_names] %>%
  map_df(read_delim, delim = "\t", col_types = 'cccddd', .id = "sample_k") %>% mutate(across(sample_k, ~str_remove(., "_minmap")))


```

Use when only running hisat2
```{r, eval=FALSE}
soupc_all <- list.files(paste0("data/processed/", species), pattern = "clusters.tsv", full.names = T, recursive = T, include.dirs = T)%>%
  set_names(nm = (str_remove_all(.,paste0('data/processed/', species, '/|soupc/h.*k|/k|/clusters.tsv')) %>% str_replace_all(., '/', '_'))) %>% .[sorted_names] %>%
  map_df(read_delim, delim = "\t", col_types = 'cccddd', .id = "sample_k") %>% mutate(across(sample_k, ~str_remove(., "_hsat")))


```

Use when only running several aligners

```{r, eval=T}
soupc_all_al <- list.files(paste0("data/processed/", species), pattern = "clusters.tsv", full.names = T, recursive = T, include.dirs = T)%>% set_names(nm = (str_remove_all(.,paste0('data/processed/', species, '/|soupc/|k|/clusters.tsv'))) %>% str_replace_all(., '/', '_'))  %>% .[sorted_names_al]%>%
  map_df(read_delim, delim = "\t", col_types = 'cccddd', .id = "sample_k") 


```


Select only information from clusters.tsv for the best k as determined from the log likelihood plot for each MSC sample

```{r, eval=T}
##Change Strain to SC0,SC1 for publication figures and enhanced readability
# soupo_kselect <- soupc_all %>%
soupo_kselect_al <- soupc_all_al %>%
  # filter(sample_k %in% knee_k) %>%
  dplyr::select(sample_k, barcode, status, assignment) %>% 
  dplyr::mutate("Strain" = case_when(status == "doublet" ~ "Doublet",
                                     status == "unassigned" ~ "Negative",
                                     assignment == "0" ~ "SC1",
                                     assignment == "1" ~ "SC2",
                                     assignment == "2" ~ "SC3",
                                     assignment == "3" ~ "SC4",
                                     assignment == "4" ~ "SC5",
                                     assignment == "5" ~ "SC6",
                                     assignment == "6" ~ "SC7",
                                     assignment == "7" ~ "SC8",
                                     assignment == "8" ~ "SC9",
                                     assignment == "9" ~ "SC10",
                                     assignment == "10" ~ "SC11",
                                     assignment == "11" ~ "SC12",
                                     assignment == "12" ~ "SC13",
                                     assignment == "13" ~ "SC14",
                                     assignment == "14" ~ "SC15",
                                     assignment == "15" ~ "SC16")
  ) 
```




```{r, eval=T}
# Compare HISAT and MINIMAP
soupo_kselect_al_cor <- soupo_kselect_al %>%
  # separate_wider_regex(cols = "sample_k", c(sample = "MSC[0-9]*", "_",aligner = "[a-z].*", "_", k = "\\d+")) %>%
  separate_wider_regex(cols = "sample_k", c(sample = "MSC[0-9]*|MSC[0-9]*_[A-Z]*", "_",aligner = "[a-z].*", "_", k = "\\d+")) %>%
  unite(col = "sample_k", c('sample','k'), remove = FALSE)  %>%
  select("sample_k","aligner","barcode","Strain") %>% 
  dplyr::group_split(sample_k) %>%
  purrr::set_names(purrr::map_chr(., ~.x$sample_k[1])) %>%
  .[sorted_names]


# imap(soupo_kselect_al_cor, ~pivot_wider(.x, names_from = aligner, values_from = Strain) %>% count(minmap, hsat) %>% pivot_wider(names_from = hsat, values_from = n) %>% column_to_rownames('minmap') %>% as.matrix() %>% ComplexHeatmap::pheatmap(., cluster_rows = F, cluster_cols = F, display_numbers = T, main = .y, annotation_names_col = T))
   
hsat_mnmp_comp = map(soupo_kselect_al_cor, ~pivot_wider(.x, names_from = aligner, values_from = Strain) %>%
  dplyr::select(barcode, all_of(algn_nms)))
  
# Compare HISAT and LR_REF
# hsat_lr_ref_comp = map(soupo_kselect_al_cor, ~pivot_wider(.x, names_from = aligner, values_from = Strain) %>%
#   dplyr::select(barcode, `lr_ref`, `hsat`))

# Compare MINIMAP and LR_REF
# mnmp_lr_ref_comp = map(soupo_kselect_al_cor, ~pivot_wider(.x, names_from = aligner, values_from = Strain) %>%
#   dplyr::select(barcode, `lr_ref`, `minmap`))
```

### Assess concordance between mapping methods

Function to plot heatmaps to assess concordance
```{r , warning=FALSE}
## Heatmap function for plotting pheatmap for number of cells assigned labels from Clara's and Roser's datasets
phmap_lab_transfr <- function(dset_lab, col_labs, row_labs, row_labs_c, col_labs_c, strn_order, ttl){
  anno_row = dset_lab %>% dplyr::count({{ col_labs }}, {{ row_labs }}) %>% 
    pivot_wider(values_from = n, names_from = {{ col_labs }}) %>% 
    column_to_rownames(names(.)[1]) %>% 
    mutate({{ row_labs }} := row_labs_c) %>%
    dplyr::select({{ row_labs }})
  
  anno_col = dset_lab %>% dplyr::count({{ col_labs }}, {{ row_labs }}) %>%
    pivot_wider(values_from = n, names_from = {{ row_labs }}) %>%
    column_to_rownames(names(.)[1]) %>%
    mutate({{ col_labs }} := col_labs_c) %>%
    dplyr::select({{ col_labs }})
  
  dset_lab %>% dplyr::count({{ col_labs }}, {{ row_labs }}) %>%
    arrange({{ col_labs }}, {{ row_labs }})%>%
    pivot_wider(values_from = n, names_from = {{ col_labs }}) %>%
    replace(is.na(.), 0) %>%
    column_to_rownames(names(.)[1])%>%
    # .[,strn_order] %>%
    ComplexHeatmap::pheatmap(display_numbers = T,treeheight_row=0, treeheight_col=0, annotation_row = anno_row, annotation_col = anno_col, legend = F, annotation_legend = T, cluster_rows =F, cluster_cols = F, fontsize = 20, annotation_names_row=T, annotation_names_col = T, 
                       # color = 'white',
                       number_format = "%.0f", main = ttl#,
                       # annotation_colors = list('minmap' = "green", 'hsat' = "lightblue")
                       )

}


```

#### HISAT VS MINIMAP {.tabset} 
```{r , warning=FALSE}

pht_mnmp_comp <- imap(hsat_mnmp_comp, possibly(~phmap_lab_transfr(.x, col_labs= `minmap`, row_labs=`hsat`, row_labs_c ='minmap', col_labs_c= 'hsat', ttl = .y)), 'k=1 so no groups to compare which is necessary for the heatmap')

```

```{r, fig.width = 7, fig.height = 4, results='asis'}

iwalk(pht_mnmp_comp,~{
        cat('##### ',.y,'   \n')
        print(.x)
        cat('\n \n')
      })
```

#### HISAT VS LR_REF


```{r , fig.width = 10, fig.height = 5, eval = F}
## Heatmap - Compare HISAT and LR_REF
imap(hsat_lr_ref_comp, possibly(~phmap_lab_transfr(.x, col_labs= `lr_ref`, row_labs=`hsat`, row_labs_c ='lr_ref', col_labs_c= 'hsat', ttl = .y)), 'k=1 so no groups to compare which is necessary for the heatmap')

```

#### MINIMAP VS LR_REF

```{r , fig.width = 10, fig.height = 5, eval = F}
## Heatmap - Compare MINIMAP and LR_REF
imap(mnmp_lr_ref_comp, possibly(~phmap_lab_transfr(.x, col_labs= `lr_ref`, row_labs=`minmap`, row_labs_c ='lr_ref', col_labs_c= 'minmap', ttl = .y)), 'k=1 so no groups to compare which is necessary for the heatmap')

```

list("MSC27" = c(2), "MSC35" = c(2:4),  "MSC36" = c(2:3), "MSC42" = c(2), "MSC44" = c(2:4), "MSC47" = c(2:4), "MSC16" = c(2:3), "MSC15" = c(2:4))

### Preprocess and put strain assignments into a list


Split the information for each sample into different labels and put into a list named appropriately
```{r}
soupo_kselect_list <- soupo_kselect_al %>%
# filter(str_detect(sample_k, algn_nm)) %>%
# soupo_kselect_list <- soupo_kselect %>%
  # mutate(across(sample_k, ~str_remove(.,paste0("_", algn_nm)))) %>%
  dplyr::group_split(sample_k) %>%
  purrr::set_names(purrr::map_chr(., ~.x$sample_k[1])) %>%
  map(. %>% column_to_rownames(., "barcode") %>% dplyr::select(Strain))

soupo_kselect_list <- soupo_kselect_list[as.character(sorted_names_al)]


##Get strain levels for samples with more than one strain and mutate Strain column to reflect this
# strain_levels <- map(soupo_kselect_list[2:8], ~.x %>% filter(str_detect(Strain, 'Strain')) %>% pull(Strain) %>% unique() %>% sort() %>% c(.,"Doublet","Negative"))
strain_levels <- map(soupo_kselect_list, ~.x %>% dplyr::filter(str_detect(Strain, '^SC')) %>% pull(Strain) %>% unique() %>% sort() %>% c(.,"Doublet","Negative"))

soupo_kselect_list <- map2(soupo_kselect_list,strain_levels, ~.x %>% mutate(across(Strain, ~factor(., levels = .y)))) 
```



```{r, message=FALSE}
##Copy souporcell files for the optimum clusters investigation on igv - all except negative

soupo_kmerge <- map(sample_name_nms, ~{
soupo_kselect_list[str_detect(names(soupo_kselect_list), paste0(.x, "_"))] %>% 
  imap(., ~{
    nms = .y
    .x %>% rename_with(~paste0(.,'_', nms), everything()) %>% rownames_to_column('bcode')
        }) %>% 
    purrr::reduce(., left_join, by = 'bcode')
  }) %>% 
  set_names(sample_name_nms)

```


```{r, message=FALSE}
##Copy souporcell files for the optimum clusters investigation on igv - all except negative

soupo_kmerge <- map(sm_aln_nms_al, ~{
soupo_kselect_list[str_detect(names(soupo_kselect_list), paste0(.x, "_"))] %>% 
  imap(., ~{
    nms = .y
    .x %>% rename_with(~paste0(.,'_', nms), everything()) %>% rownames_to_column('bcode')
        }) %>% 
    purrr::reduce(., left_join, by = 'bcode')
  }) %>% 
  set_names(sm_aln_nms_al)

```

## Harmonize strain/cluster names across a range of Ks for each donor

Function to harmonize the strain names so that the same SC cluster can be traced from k=1 onwards
```{r, fig.width=4, fig.height=4}


strain_harmonize_fn <- function(strn_tbl = tbl1, strn1 = "Strain_2nd", strn2 = "Strain_final", out_strn = "n_strains") {
  
  strn_tbl %>% group_by(!!rlang::sym(strn1)) %>% 
            add_count(!!rlang::sym(strn2), name = "stf_n") %>%
            mutate(across(stf_n, ~case_when(str_detect(!!rlang::sym(strn2), "SC") & str_detect(!!rlang::sym(strn1), "Doublet|Negative") ~ 0,
                                            str_detect(!!rlang::sym(strn1), "SC") & str_detect(!!rlang::sym(strn2), "Doublet|Negative") ~ 0,
                                            str_detect(!!rlang::sym(strn1), "Doublet|Negative") & str_detect(!!rlang::sym(strn2), "Doublet|Negative") ~ 0,
                                            T ~ .)),
                   top_st = max(stf_n),
                   nw_st = case_when(!!rlang::sym(strn2) == "Doublet" | !!rlang::sym(strn2) == "Negative" ~ !!rlang::sym(strn2),
                                     top_st == stf_n & top_st != 0 ~ !!rlang::sym(strn1))) %>% 
            group_by(top_st, nw_st) %>%
            mutate(lab_uniq = row_number(),
                   across(lab_uniq, ~case_when(top_st == 0 | nw_st == "Doublet" | nw_st == "Negative" ~ NA, T ~ .))) %>%
            group_by(!!rlang::sym(strn2)) %>%
            mutate(top_st2 = case_when(top_st == stf_n & top_st != 0 & lab_uniq == 1 ~ top_st, T ~ 0), 
                   top_top_st = max(top_st2),
                   ty_brakr = case_when(!!rlang::sym(strn2) == "Doublet" | !!rlang::sym(strn2) == "Negative" ~ !!rlang::sym(strn2),
                                        top_top_st == stf_n & top_top_st != 0 ~ !!rlang::sym(strn1))) %>% 
        arrange(!!rlang::sym(strn2))  %>%
        fill(ty_brakr, .direction = "downup") %>%
        group_by(is.na(ty_brakr))%>%
        mutate(nro_count = data.table::rleid(!!rlang::sym(strn2))) %>% ## Gets the count of distinct items in !!rlang::sym(strn2) into row numbers within the 'is.na(ty_brakr)' groups
        ungroup()%>%
        mutate(sc_numbr = as.numeric(str_remove_all(!!rlang::sym(strn1), "Doublet|Negative|SC")))%>%
        mutate(!! out_strn:= case_when(is.na(ty_brakr) ~ paste0("SC",(max(sc_numbr, na.rm = T)+nro_count)), T ~ ty_brakr)) %>%
        # select(c("bcode", strn1, out_strn)) %>% 
        select(-c('stf_n', 'top_st',   'ty_brakr', 'is.na(ty_brakr)', 'nro_count')) 
  
}
```

### Visualize the transition of strains/clusters from one K to the next

Sankey plotting function
```{r, fig.width=4, fig.height=4}

snk_pt_fn <- function(decod_tbl, strns_2_plt = c('Strain_2nd', 'n_strains'), labls = paste0("K=", opt_clusters_nms[[3]][1:2]), bp_face = c("bold", "plain"), ttl = "plot"){
  decod_tbl %>% 
        mutate(across(contains("train"), ~str_replace_all(., c("Doublet" = "Dbt", "Negative" = "Neg")))) %>%
        make_long(all_of(strns_2_plt)) %>% 
        add_count(x,node, "n") %>%
        ggplot(., aes(x = x, next_x = next_x,node = node,next_node = next_node,fill = factor(node),label = paste0(node,": ",n))) +
            geom_sankey() +
        theme_void()+
        scale_fill_manual(values = strain_cols_n)+
        scale_x_discrete(labels=labls)+
        geom_sankey_label(size = 3.6, color = 1, fill = "white") +
        labs(title = ttl) +
        theme_sankey(base_size = 9) +
        theme(legend.position = "none",
              axis.title.x = element_blank(),
              axis.text.x = element_text(size = 15, face = bp_face))
      
}

```


```{r, fig.width=4, fig.height=4}

don_k_names_aln <- imap(opt_clusters_nms_aln, ~paste0('Strain_',.y, '_', .x))
don_hmonizd_k_names_aln <- imap(opt_clusters_nms_aln, ~paste0('Strain_hmo_',str_remove(.y, "SC"), '_', .x))

# # strn_harmnzd_tbls <- list(soupo_kmerge, don_k_names, don_hmonizd_k_names) %>%
# strn_harmnzd_tbls <- list(soupo_kmerge, don_k_names_aln, don_hmonizd_k_names_aln) %>%
#   pmap(~{
#     
#     strain_harmonize_fn(strn_tbl = ..1, strn1 = ..2[1], strn2 = ..2[2], out_strn = ..3[2]) %>%
#       strain_harmonize_fn(strn_tbl = ., strn1 = ..3[2], strn2 = ..2[3], out_strn = ..3[3])
#    
#   })
#     
# # list(strn_harmnzd_tbls, don_k_names, opt_clusters_nms, don_hmonizd_k_names) %>%
# list(strn_harmnzd_tbls, don_k_names_aln, opt_clusters_nms_aln, don_hmonizd_k_names_aln) %>%
#   pmap(~{  
#     
# snk_pt_plts <- snk_pt_fn(..1, strns_2_plt = c(..2[1], ..4[2:3]), labls = paste0("K=", ..3[1:3]), bp_face = c("bold", "plain", "plain"))
#     
#  })

```


Funtion to create the input for the sankey plotting function - limited to 5 strains can be altered to add more
```{r, fig.width=6, fig.height=4}

strn_hm_tbl_fn <- function(soupo_seu_kmerge_f, don_k_names_aln_f, don_hmonizd_k_names_aln_f){
  list(soupo_seu_kmerge_f, don_k_names_aln_f, don_hmonizd_k_names_aln_f) %>%
    pmap(~{
      
      if (length(..2) == 3) { 
          
            strain_harmonize_fn(strn_tbl = ..1, strn1 = ..2[1], strn2 = ..2[2], out_strn = ..3[2]) %>%
            strain_harmonize_fn(strn_tbl = ., strn1 = ..3[2], strn2 = ..2[3], out_strn = ..3[3])
         
          } else if (length(..2) == 4) {
          
            strain_harmonize_fn(strn_tbl = ..1, strn1 = ..2[1], strn2 = ..2[2], out_strn = ..3[2]) %>%
            strain_harmonize_fn(strn_tbl = ., strn1 = ..3[2], strn2 = ..2[3], out_strn = ..3[3])%>%
            strain_harmonize_fn(strn_tbl = ., strn1 = ..3[3], strn2 = ..2[4], out_strn = ..3[4])
       
          } else if  (length(..2) == 5) {
          
            strain_harmonize_fn(strn_tbl = ..1, strn1 = ..2[1], strn2 = ..2[2], out_strn = ..3[2]) %>%
            strain_harmonize_fn(strn_tbl = ., strn1 = ..3[2], strn2 = ..2[3], out_strn = ..3[3])%>%
            strain_harmonize_fn(strn_tbl = ., strn1 = ..3[3], strn2 = ..2[4], out_strn = ..3[4])%>%
            strain_harmonize_fn(strn_tbl = ., strn1 = ..3[4], strn2 = ..2[5], out_strn = ..3[5])
         
          } else {
          print("K range (number of strain iterations) is more than 5. Limit to 5")
      }
    })
}


```



```{r, fig.width=6, fig.height=4}

strn_harmnzd_tbls <- strn_hm_tbl_fn(soupo_seu_kmerge_f = soupo_kmerge, don_k_names_aln_f =don_k_names_aln, don_hmonizd_k_names_aln_f = don_hmonizd_k_names_aln)
```

```{r, fig.width=6, fig.height=4, eval=T}
# list(strn_harmnzd_tbls_5, don_k_names, opt_clusters_nms, don_hmonizd_k_names) %>%
snk_plts <- list(strn_harmnzd_tbls, don_k_names_aln, opt_clusters_nms_aln, don_hmonizd_k_names_aln) %>%
  pmap(~{  
    
snk_pt_fn(..1, strns_2_plt = c(..2[1], ..4[2:length(..4)]), labls = paste0("K=", ..3[1:5]), bp_face = c("plain"))
    
 })

```

#### Minimap sankey plots  {.tabset} 
```{r, eval=T}
snk_plts_minmap <- snk_plts[str_detect(names(snk_plts), "minmap")] %>% imap(., ~.x + labs(title = .y))
```

```{r, results='asis'}

iwalk(snk_plts_minmap,~{
        cat('##### ',.y,'   \n')
        print(.x)
        cat('\n \n')
      })
```

#### Hisat sankey plots {.tabset} 
```{r, eval=T}
snk_plts_hsat <- snk_plts[str_detect(names(snk_plts), "hsat")] %>% imap(., ~.x + labs(title = .y))
```

```{r, results='asis'}

iwalk(snk_plts_hsat,~{
        cat('##### ',.y,'   \n')
        print(.x)
        cat('\n \n')
      })
```

#### lr_ref sankey plots
```{r, eval=FALSE}
snk_plts[str_detect(names(snk_plts), "lr_ref")]
```


```{r}
## Choose optimum k based on the above results
optimum_k = c("MSC27" = 1, "MSC35" = 3, "MSC42" = 1, "MSC44" = 3, "MSC47" = 6, "MSC36" = 2, "MSC16" = 2, "MSC15" = 3) %>% .[sample_name_nms]

## having k=2 for samples with k=1 to verify no additional strains

optimum_k_aln = rep(optimum_k, length(algn_nms)) %>% set_names(sm_aln_nms_al)
```

### Get harmonized tables and assess whether done correctly
Important as the strain categories may be fumbled up

Get harmonized strain names table

```{r, fig.width=6, fig.height=4}
## Rename the initial k colnames which is unchanged when harmonizing strain annotation for Sankey plot and factorized newly generated strain harmonized vectors
# strn_harmnzd_tbls_5 <- list(strn_harmnzd_tbls_5, map(don_k_names, ~.x[1]), map(don_hmonizd_k_names, ~.x[1])) %>%
strn_harmnzd_tbls <- list(strn_harmnzd_tbls, map(don_k_names_aln, ~.x[1]), map(don_hmonizd_k_names_aln, ~.x[1])) %>%
  pmap(~{
    ..1 %>% 
      mutate(!! ..3:= !!rlang::sym(..2)) %>%
      mutate(across(contains("Strain_hmo"), ~droplevels(factor(.,levels = strain_levels_univrs))))
    
  })

```

Check harmonization of strain names
```{r, results='hide'}
## !! NOTE - VERY IMPORTANT to verify that the strain renaming to harmonize names for sankey plot function has been done correctly
# strn_harmnzd_assesmnt <- list(strn_harmnzd_tbls_5, don_k_names, don_hmonizd_k_names) %>%
strn_harmnzd_assesmnt <- list(strn_harmnzd_tbls, don_k_names_aln, don_hmonizd_k_names_aln) %>%
  pmap(~{
    tbl <- ..1
    nms <- ..2
    nms_hmo <- ..3
    
    map2(nms, nms_hmo,  ~{
      tbl %>% 
        mutate("strn_o" =!!rlang::sym(.x) , "strn_n" = !!rlang::sym(.y)) %>% 
        dplyr::count(strn_o, strn_n) %>% 
        arrange(strn_o)
      }) %>% 
      set_names(nms) %>% 
      bind_rows(., .id = "strn_k") 
    
  })

strn_harmnzd_assesmnt
```

```{r, fig.width=6, fig.height=4}
## !! NOTE - VERY IMPORTANT to verify that the strain renaming to harmonize names for sankey plot function has been done correctly - Detailed assesment - may be time consuming for many donor samples
# bind_rows(strn_harmnzd_assesmnt) %>% view
```

```{r, results='hide'}
## !! NOTE - VERY IMPORTANT to verify that the strain renaming to harmonize names for sankey plot function has been done correctly - Summary of table above
hmonizd_k_test_summary <- bind_rows(strn_harmnzd_assesmnt) %>% group_by(strn_k) %>% mutate(dup = duplicated(strn_o) | duplicated(strn_n)) %>% count(dup) %>% mutate(k=n-2) %>% mutate(k_test = as.numeric(str_remove(strn_k, "Str.*_")) == k)
hmonizd_k_test_summary

```

```{r, fig.width=6, fig.height=4}
## !! NOTE - VERY IMPORTANT to verify that the strain renaming to harmonize names for sankey plot function has been done correctly - Verify those that are FALSE
table(hmonizd_k_test_summary$k_test)
```


```{r, fig.width=4, fig.height=4, eval=FALSE}
## !!NOTE DON'T DELETE - For troubleshooting functions above

# soupo_kmerge[[3]] %>% 
#             mutate("Strain_final" = !!rlang::sym(paste0("Strain_","MSC49","_", 7)),
#                    "Strain_2nd" = !!rlang::sym(paste0("Strain_","MSC49","_", 6))) %>%
#             select(bcode, Strain_final, Strain_2nd) %>% 
#             group_by(Strain_2nd) %>% 
#             count(Strain_final, name = "stf_n") %>%
#             mutate(across(stf_n, ~case_when(str_detect(Strain_final, "SC") & str_detect(Strain_2nd, "Doublet|Negative") ~ 0,
#                                             str_detect(Strain_2nd, "SC") & str_detect(Strain_final, "Doublet|Negative") ~ 0,
#                                             str_detect(Strain_2nd, "Doublet|Negative") & str_detect(Strain_final, "Doublet|Negative") ~ 0,
#                                             T ~ .)),
#                    top_st = max(stf_n),
#                    nw_st = case_when(Strain_final == "Doublet" | Strain_final == "Negative" ~ Strain_final,
#                                      top_st == stf_n & top_st != 0 ~ Strain_2nd)) %>% 
#             group_by(top_st, nw_st) %>%
#             mutate(lab_uniq = row_number(),
#                    across(lab_uniq, ~case_when(top_st == 0 | nw_st == "Doublet" | nw_st == "Negative" ~ NA, T ~ .))) %>%
#             group_by(Strain_final) %>%
#             mutate(top_st2 = case_when(top_st == stf_n & top_st != 0 & lab_uniq == 1 ~ top_st, T ~ 0),
#                    top_top_st = max(top_st2),
#                    ty_brakr = case_when(Strain_final == "Doublet" | Strain_final == "Negative" ~ Strain_final,
#                                         top_top_st == stf_n & top_top_st != 0 ~ Strain_2nd))%>% 
#                                     #top_st == stf_n & top_st ~ Strain_2nd)) %>% 
#         arrange(Strain_final)  %>%
#         fill(ty_brakr, .direction = "downup") %>%
#         group_by(is.na(ty_brakr))%>%
#         mutate(nro_count = data.table::rleid(Strain_final)) %>% ## Gets the count of distinct items in Strain_final into row numbers within the 'is.na(ty_brakr)' groups
#         ungroup()%>%
#         mutate(max_sc = as.numeric(str_remove_all(Strain_2nd, "Doublet|Negative|SC")))%>%
#         mutate(n_strains= case_when(is.na(ty_brakr) ~ paste0("SC",(max(max_sc, na.rm = T)+nro_count)), T ~ ty_brakr))%>%
#   view

```

## Narrow down to smaller range (max = 3) of K for investigation by pseudobulk genotyping

```{r, message=FALSE}
##Coalesce the genotypes and get a consensus genotype taking labels for highest resolution K
##VARIABLE
## !!!NB!!!!!When trialing out different souporcell Ks use Strain_qc_cutoff > 20 but once decided use Strain_qc_cutoff > 0 the optimum K 

# Strain_qc_cutoff=20
Strain_qc_cutoff=0
# optimum_k = c(8,5,6,7,7) %>% set_names(sample_name_nms)

opt_narrow_clusters_nms = list("MSC27" = 2, "MSC35" = c(2:4), "MSC42" = 2, "MSC44" = c(2:4), "MSC47" = c(5:7), "MSC36" = c(2:3), "MSC16" = c(2:3), "MSC15" = c(2:4)) %>% .[sample_name_nms]

opt_narrow_clusters_nms_aln <- rep(opt_narrow_clusters_nms, length(algn_nms)) %>% set_names(sm_aln_nms_al)

```


```{r, message=FALSE}
##Coalesce the genotypes and get a consensus genotype taking labels for highest resolution K
##VARIABLE
## !!!NB!!!!!When trialing out different souporcell Ks use Strain_qc_cutoff > 20 but once decided use Strain_qc_cutoff > 0 the optimum K 

# strn_c_tbl <- list(strn_harmnzd_tbls_5, names(strn_harmnzd_tbls_5), opt_clusters_nms[names(strn_harmnzd_tbls_5)] %>% map(.,max), optimum_k, imap(opt_narrow_clusters_nms, ~paste0(.y, '_', .x))) %>%
strn_c_tbl <- list(strn_harmnzd_tbls, 
                     names(strn_harmnzd_tbls), 
                     opt_clusters_nms_aln[names(strn_harmnzd_tbls)] %>% map(.,max), 
                     optimum_k_aln, 
                     imap(opt_narrow_clusters_nms_aln, ~paste0(.y, '_', .x))) %>%
  pmap(~{
    ..1 %>%
      unite(col = 'Strn_c', starts_with("Strain_hmo_M"), sep='_', remove = F) %>% 
      add_count(Strn_c) %>% 
      mutate(
        'Strain_max_qc' = case_when( 
          !str_detect(Strn_c, 'Negative|Doublet') & n>Strain_qc_cutoff ~ !!rlang::sym(paste0('Strain_hmo_', str_remove(..2, 'SC'), '_',..3)), 
          T ~ 'PoorQC'),
        'Strain_qc' = case_when(
          !str_detect(!!rlang::sym(paste0('Strain_hmo_', str_remove(..2, 'SC'), '_',..4)), 'Negative|Doublet') & n>Strain_qc_cutoff ~ !!rlang::sym(paste0('Strain_hmo_', str_remove(..2, 'SC'), '_',..4)), 
          T ~ 'PoorQC'),
        'Strain_DN' = !!rlang::sym(paste0('Strain_hmo_', str_remove(..2, 'SC'), '_',..4)),
        'dbt_neg_all' = case_when(
          if_all(c(paste0('Strain_', ..5)), ~ .== "Doublet") ~ "DoubletAll",
          if_all(c(paste0('Strain_', ..5)), ~ .== "Negative") ~ "NegativeAll",
          if_all(c(paste0('Strain_', ..5)), ~ .== "Doublet" | .== "Negative") ~ "DbtNegMix", 
          T ~ 'LikelySinglet')
        )%>%
        dplyr::add_count(Strain_qc, name = 'cluster_n') 
  })



```


```{r}
##Save strain + new metadata as RDS
map(strn_c_tbl, 
    . %>% 
      # dplyr::select("bcode", "Strn_c", starts_with("Strain"),"stage_afm", "stage_ag", "cluster_ag_n", "cluster_afm_n", "stageHL", "strain_afm_n", "strain")) %>%
      dplyr::select(-c(starts_with("Strain_MSC"))) %>%
      dplyr::select("bcode", "Strn_c", starts_with("Strain"), "dbt_neg_all")) %>%
  saveRDS(paste0("data/processed/", species,"/strn_c_tbl.RDS"))


```

Write out strain barcodes to farm for pseudobulk genotyping


```{r, eval=T, results='hide'}
## DOUBLETS WRITTEN INTO THE SAME FOLDER IN MSC_PP.rmd - SHOULD BE COMBINED INTO ONE NOTEBOOK
# sample_ids = map2(sample_name_nms, source_irods_nms, ~paste0(.x, '_', .y))

for (s in names(strn_c_tbl)) {
  narrow_k <- opt_narrow_clusters_nms[[str_remove(s, "_[a-z]+")]]
  ## Assign minimap/hsat mapper directory
  algn_nm = str_extract(s, "hsat|minmap|lr_ref")
  s_nm = str_remove(s, "_[a-z]+")
  
  print(algn_nm)
  print(s_nm)
  
  ## !!!NOTE - Dangerous command - activate only when barcodes are modified
  
  system(paste0('rm -r data/processed/',species,'/',s_nm,'/', sv_dir,'/bcodes/',algn_nm,'/st*'))
  
  for (k in narrow_k) {
    for (str_q in c('strain')){
      
      ss_qcd_ano <- map(strn_c_tbl[s], ~.x %>% dplyr::select(paste0('Strain_hmo_', str_remove(s_nm, 'SC'), '_', algn_nm, '_',k), bcode) %>% rename('str'= !!rlang::sym(paste0('Strain_hmo_', str_remove(s_nm, 'SC'), '_', algn_nm, '_',k))) %>% group_by(str) %>% filter(n() >= 20) %>% ungroup() %>% mutate(across(c(str), ~droplevels(factor(.)))) %>% split(f = ~ str, drop = T))
      names(ss_qcd_ano[[1]]) <- names(ss_qcd_ano[[1]]) %>% str_replace(., '\\.', '_')
      print(names(ss_qcd_ano[[1]]))

      system(paste0('mkdir -p data/processed/',species,'/',s_nm,'/', sv_dir,'/bcodes/', algn_nm,'/',str_q,'_k',k,'/'))
      print(paste0('mkdir -p data/processed/',species,'/',s_nm,'/', sv_dir,'/bcodes/', algn_nm,'/',str_q,'_k',k,'/'))

      imap(ss_qcd_ano[[1]], ~write(.x$bcode, paste0('data/processed/',species,'/',s_nm,'/', sv_dir,'/bcodes/',algn_nm,'/',str_q,'_k',k,'/',.y, '.tsv')))


    }
      system(paste0('mkdir -p data/processed/',species,'/',s_nm,'/', sv_dir,'/bcodes/',algn_nm, '/bcodes_no_dbt' ))

      strn_c_tbl[[s]] %>% filter(dbt_neg_all == "LikelySinglet") %>% pull(bcode) %>% write(., paste0('data/processed/',species,'/', s_nm, '/', sv_dir,'/bcodes/',algn_nm, '/bcodes_no_dbt/barcodes.tsv'))

  }

}
```



```{r, eval=FALSE}
gogogo()

# nextflow run pbulk_gtyping.nf --sv_dir "pbulk_gtypes_preQC" --bam "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pm/MSC*/soupc/*/parent/possorted_genome_bam.bam"  --bcodes "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pm/MSC*/pbulk_gtypes_preQC/bcodes/minmap/*/*tsv" --odir "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/data/processed/Pm/" --mapper "minmap" --scrpt "/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/multipurpose_scripts/cr_subset_bam_linux.sh" --ref  "/lustre/scratch126/tol/teams/lawniczak/projects/malaria_single_cell/mali_field_runs/2022/data/references_for_souporcell/PlasmoDB-66_PmalariaeUG01_Genome.fasta" -resume

```


## PCA strain visualization
```{r}
# soupo_PC_mtx <- soupc_all %>% 
soupo_PC_mtx <- soupc_all_al %>% 
  dplyr::select(-c(status, assignment, log_prob_singleton, log_prob_doublet)) %>% 
  dplyr::group_split(sample_k) %>%
  purrr::set_names(purrr::map_chr(., ~.x$sample_k[1])) %>%
  map(. %>% column_to_rownames(., "barcode") %>%
        dplyr::select(where(~!all(is.na(.)))) %>%
        dplyr::select(-sample_k) %>%
        mutate(across(where(is.numeric), ~ (. / rowSums(across(where(is.numeric)))))) %>%
        as.matrix() %>%
        t())

# soupo_PC_mtx <- soupo_PC_mtx[as.character(sorted_names)]
soupo_PC_mtx <- soupo_PC_mtx[as.character(sorted_names_al)]
```



```{r}
##Calculate PCA
soupo_PCA <- soupo_PC_mtx %>% 
  map(. %>% prcomp(.))
```


```{r}
## Save PCA
saveRDS(soupo_PCA, paste0("data/processed/", species,"/soupo_PCA_m22.RDS"))
```

Calculate proportion of variance explained by PC1 and PC2
```{r}
# Assuming your data is stored in a matrix called 'data'
# Extract the standard deviations of the principal components
# pc_standard_deviations <- soupo_PCA[[9]]$sdev
# 
# #PC1 variance explained
# (pc_standard_deviations[1] ^ 2) / sum(pc_standard_deviations^2)
# 
# #PC2 variance explained
# (pc_standard_deviations[2] ^ 2) / sum(pc_standard_deviations^2)
```


```{r, results='hide'}
sm_aln_opt_ks_nms <- unlist(map2(names(strn_c_tbl), rep(opt_clusters_len, length(algn_nms)), ~rep(.x, each = .y)))
strn_c_tbl_sm_aln_opt_ks <- unlist(map2(strn_c_tbl, rep(opt_clusters_len, length(algn_nms)), ~rep(list(.x), times = .y)), recursive = F)


list(soupo_PCA, strn_c_tbl_sm_aln_opt_ks, names(soupo_PCA)) %>%
  pmap(~{..1$rotation %>% 
     as.data.frame() %>%
     rownames_to_column('bcode') %>% 
      left_join(..2, by = 'bcode') %>%
     head
      })
```

### PC1_PC2

```{r}
# list(soupo_PCA, rep(strn_c_tbl, each = 5, times = 2), names(soupo_PCA)) %>%
strn_pc_lts <- list(soupo_PCA, strn_c_tbl_sm_aln_opt_ks, names(soupo_PCA)) %>%
  pmap(~{..1$rotation %>% 
     as.data.frame() %>%
     rownames_to_column('bcode') %>% left_join(..2, by = 'bcode') %>%
     ggplot(., aes(x=PC1, y=PC2, color = Strain_qc)) + 
     geom_point(size=2) +
    labs(title = ..3) +
      theme_classic()
      })
```

#### Minimap {.tabset} 
```{r, results='asis'}
strn_pc_lts[str_detect(names(strn_pc_lts), "minmap")] %>% 
      purrr::iwalk(.,~{
        cat('##### ',.y,'   \n')
        print(.x)
        cat('\n \n')
      })
```

#### Hisat {.tabset} 
```{r, results='asis'}
strn_pc_lts[str_detect(names(strn_pc_lts), "hsat")] %>% 
      purrr::iwalk(.,~{
        cat('##### ',.y,'   \n')
        print(.x)
        cat('\n \n')
      })
```

### PC1_PC3
```{r}
strn_pc1_3_lts <- list(soupo_PCA, strn_c_tbl_sm_aln_opt_ks, names(soupo_PCA)) %>%
  pmap(~{..1$rotation %>% 
     as.data.frame() %>%
     rownames_to_column('bcode') %>% left_join(..2, by = 'bcode') %>%
     ggplot(., aes(x=PC1, y=PC3, color = Strain_qc)) + 
     geom_point(size=2)+
    labs(title = ..3) +
      theme_classic()
    })


```

#### Minimap {.tabset} 
```{r, results='asis'}

cat('\n')
strn_pc1_3_lts[str_detect(names(strn_pc1_3_lts), "minmap")] %>% 
      purrr::iwalk(.,~{
        cat('##### ',.y,'   \n')
        print(.x)
        cat('\n \n')
      })
```

#### Hisat {.tabset} 
```{r, results='asis'}

cat('\n')
strn_pc1_3_lts[str_detect(names(strn_pc1_3_lts), "hsat")] %>% 
      purrr::iwalk(.,~{
        cat('##### ',.y,'   \n')
        print(.x)
        cat('\n \n')
      })
```
