---
title: "Pm strain deconvolution for pre and post QC each donor singly"
date: '2024-07-31'
author: 'Jesse Rop'
output:
  html_document:
    toc: true
    theme: united
    code_folding: hide
    df_print: paged
---


```{r, message=FALSE}
suppressPackageStartupMessages({
  # library(Seurat)
  library(dplyr) 
  library(stringr)
  library(purrr)
  library(readr)
  library(tibble)
  library(ggplot2)
  library(tidyr)
  # library(cowplot)
  library(pheatmap)
  # library(rmarkdown)
  # library('vcfR')
  library(conflicted)
  library(RColorBrewer)
  # library(MetBrewer)
  library(ComplexHeatmap)
  library(SeqArray)
  library(ggrepel)
  library(ggvenn)
  # library(magick)
  # library(VariantAnnotation)
  library(gridExtra)
  library(Matrix)
  # library(UpSetR)
  library(SNPRelate)
  # library(openxlsx)
  
  conflict_prefer("select", "dplyr")
  conflict_prefer("filter", "dplyr")
  conflict_prefer("rename", "dplyr")
})
```

# MSC Pm parasites strain deconvolution, genotyping and relatedness 

## Variable declarations and setup

```{r setup}
knitr::opts_knit$set(root.dir = paste0('/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/'))
```


```{r}
## Get the necessary variable for the pseudobulking workflow
# source("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/mali22_code/local_rmd_nbooks/m2_pbulk_gtypes_preQC_header.R")
source("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/mali22_code/local_rmd_nbooks/m2_pm_pbulk_gtypes_preQC_header.R")
sample_name_nms
```


```{r, results='hide'}
## Get the necessary variables for common housekeeping
source("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/multipurpose_scripts/plotting_common_fns.R")
source("/lustre/scratch126/tol/teams/lawniczak/users/jr35/phd/Mali2/mali22_code/pf_common_vars.R")
```
## READING IN GENOTYPES AND VARTRIX 
################### READING IN GENOTYPES AND VARTRIX ###########################

VCF to GDS conversion
```{r, warning=FALSE, message=FALSE, eval=T, results='hide'}
##Read in the genotypes and annotate

gt_ss_afm_p1_bi <-pmap(list(sample_name_k, optimum_k), ~{
  sm <- ..1
  gp <- ..2
  map(grps, ~seqVCF2GDS(paste0("data/processed/",species, "/", sm ,"/", gtyp_step, "/", .x,"_k", gp,"_", algn_nm,"/p1/fb_biallelic.vcf"), paste0("data/processed/",species, "/", sm ,"/", gtyp_step, "/", .x,"_k", gp,"_", algn_nm,"/p1/fb_biallelic.gds"),  storage.option="ZIP_RA", verbose=F)
)
})

```


```{r, results="hide"}
##Merge SNP information above with annotations from plasmoDB to get gene positions etc for plotting
# dp  %>% filter_if(any(!is.na()))
gene_n_pos <- Pf3D7_genes_plasmoDB %>% 
  mutate(
    chr = str_extract(`Genomic Location (Gene)`, '.*v3'), 
    gn_start = str_remove_all(`Genomic Location (Gene)`, '.*v3:|,|\\.\\..*'), 
    gn_end = str_remove_all(`Genomic Location (Gene)`, '.*\\.\\.|,|\\(.*')
  ) %>%
  mutate(across(c(gn_start, gn_end), ~as.numeric(.))) %>% 
  dplyr::select(gene_id, chr, gn_start, gn_end, Gene) 

# write_csv(strain_top_snps, '/Users/jr35/Google Drive/My Drive/PHD_documents/malaria_phd/Mali2/data/processed/Pf/strain_top_snps.csv')

```


```{r}
## Function to read in the genotypes and annotate

gds_rd <- function(sample_id, grp, opt_k){  
  
  p1_bi <- seqOpen(paste0("data/processed/",species, "/", sample_id ,"/", gtyp_step, "/", grp, "_k", opt_k,"_", algn_nm,"/p1/fb_biallelic.gds"), allow.duplicate=T)
  print(paste0("data/processed/",species, "/", sample_id ,"/", gtyp_step, "/", grp, "_k", opt_k,"_", algn_nm,"/p1/fb_biallelic.gds"))
  
  print(head(seqGetData(p1_bi, "allele")))
  # print(head(seqGetData(p1_bi, "$dosage")))
  # print(head(seqGetData(p1_bi, "annotation/format/QR")))

  gtypes_p1_bi <- base::rbind(seqGetData(p1_bi, "allele"), seqGetData(p1_bi, "annotation/qual"),seqGetData(p1_bi, "chromosome"), seqGetData(p1_bi, "position"), seqGetData(p1_bi, "annotation/info/DP"), seqGetData(p1_bi, "annotation/filter"), seqGetData(p1_bi, "$dosage"), seqGetData(p1_bi, "annotation/format/DP"), seqGetData(p1_bi, "annotation/format/QR"), seqGetData(p1_bi, "annotation/format/QA")) %>% t()
  
  seqClose(p1_bi)
  print(gtypes_p1_bi[1:5,1:5])
  
  strn_nms <- seqVCF_SampID(paste0("data/processed/",species, "/", sample_id ,"/", gtyp_step, "/", grp, "_k", opt_k,"_", algn_nm,"/p1/fb_biallelic.vcf")) %>% str_replace_all(c("PoorQC"="P", "Low_n"="Lo", "Female"="F", "Male" = "M", "Asexual" = "A", "Gametocyte" = "G"))
  
  colnames(gtypes_p1_bi) <- c('alleles', 'quality', 'chrom', 'pos', 'depth', 'filt', strn_nms, paste0(strn_nms, rep(c('_DP','_RefN', '_AltN'), each=length(strn_nms))))

  return(gtypes_p1_bi)
  
}
```


```{r, warning=FALSE, message=FALSE}
##Read in the genotypes and annotate
# gt_ss_afm_p1_bi <- gds_rd(sample_id = "MSC48_5736STDY13782212", grp = "stage_afm_strain_k6_minimp")

gt_ss_afm_p1_bi <-pmap(list(sample_name_k, optimum_k), ~{
  sm <- ..1
  k <- ..2
  map(grps, ~gds_rd(sample_id = sm, grp = .x, opt_k=k)) %>% set_names(paste0(..1 ,'.', grps, '.k', ..2))
  
})
```

Vartrix matrix to get number of molecules supporting either reference or alternate allele - all

```{r}
##Function to read vartrix output and annotate

vtx_rd <- function(sample_id, grp, opt_k ){
  
  ##Read in the alternate allele matrix for all the stg/strn combinations and combine into a list - also shortening the names of the matrices (list items)
  alt_mtx <- list.files(paste0("data/processed/",species, "/", sample_id ,"/", gtyp_step, "/", grp, "_k", opt_k, "_", algn_nm,"/p1/vartrix_biallelic/"), pattern = "^[D|SC|G|P|N].*.mtx", full.names = T, recursive = T, include.dirs = T) %>% 
    set_names(nm = (str_remove_all(.,paste0("data/processed/",species, "/", sample_id ,"/", gtyp_step, "/", grp, "_k", opt_k, "_", algn_nm,"/p1/vartrix_biallelic/", '/|.mtx')) %>% str_replace_all(c("PoorQC"="P", "Low_n"="Lo", "Female"="F", "Male" = "M", "Asexual" = "A", "Gametocyte" = "G", "_a_" = "_A_"))  %>%  tools::file_path_sans_ext())) %>%
    map(readMM) 
  
  # print(str(alt_mtx))
  
  ##Count, for each SNP, the total alternate allele count over all cells and proportion of cell that have alternate allele - Using base R because it is faster
  alt_mtx_stats = lapply(alt_mtx, function(x) base::cbind(x, rowSums(x) , rowSums(x != 0)/ncol(x))[,c(ncol(x)+1, ncol(x)+2)])
  
  print(c("1", sample_id, grp))
  
  ##Name the matrix columns with the name of the stg/strn and suffix of snpNum (total snp count over all cell) and celProp (proportion of cell that have alternate allele)
  nmd_alt_mt<- lapply(names(alt_mtx_stats), function(x) {
       colnames(alt_mtx_stats[[x]]) <- paste0(x, c('_sNum', '_cProp'))
       as.matrix(alt_mtx_stats[[x]])
   })
  
  # print(str(nmd_alt_mt))
  print(c("2", sample_id, grp))
  
  nmd_alt_mt <- do.call(base::cbind, nmd_alt_mt) 
  print(c("finito", sample_id, grp))
  return(nmd_alt_mt)
  
}

```


```{r}
## For troubleshooting above script
##Read in the alternate allele matrix for all the stg/strn combinations and combine into a list - also shortening the names of the matrices (list items)
sf <- list.files(paste0("data/processed/",species, "/", "MSC33" ,"/", gtyp_step, "/", "stage_afm_strain", "_k", "8", "_", algn_nm,"/p1/vartrix_biallelic/"), pattern = "^[D|SC|G|P|N].*.mtx", full.names = T, recursive = T, include.dirs = T) %>% 
    set_names(nm = (str_remove_all(.,paste0("data/processed/",species, "/", "MSC33" ,"/", gtyp_step, "/", "stage_afm_strain", "_k", "8", "_", algn_nm,"/p1/vartrix_biallelic/", '/|.mtx')) %>% str_replace_all(c("PoorQC"="P", "Low_n"="Lo", "Female"="F", "Male" = "M", "Asexual" = "A", "Gametocyte" = "G", "_a_" = "_A_"))  %>%  tools::file_path_sans_ext())) %>%
    map(readMM) 


```


```{r}
##Read in the genotypes and annotate
# NOTE - if we get error then check the header of the raw .mtx and .ref files from vartrix for consistency in number of records "number of rows of matrices must match (see arg 15)"

ss_afm_vtx_mtx <-pmap(list(sample_name_k, optimum_k), ~{
  sm <- ..1
  k <- ..2
  map(grps, ~vtx_rd(sample_id = sm, grp = .x, opt_k=k)) %>% set_names(paste0(..1 ,'.', grps))
})

```

################### END - READING IN GENOTYPES AND VARTRIX #####################

## CLEANING AND PREPROCESSING GENOTYPES
################### CLEANING AND PREPROCESSING GENOTYPES #######################

Combine vartix snp count per cell with freebayes calls
```{r}
##remove doublets as well and low quality SNPS - maybe not include doublets in variant calling with freebayes. Adds no value? or could be used to demonstrate high number of non-distinct and rule out small clusters as doublets and be part of QC

# Same as above but option to remove some groups  

fb_vt_bind <- function(fb_gtype , vtx_mtx, rm_grps="None") {
  gt_fb_vtx_p1_bi <- base::cbind(fb_gtype, vtx_mtx) %>% 
    as.data.frame() %>% 
    mutate(across(c(pos, depth, quality, matches('.*_DP$|.*_RefN$|.*_AltN$|.*sNum$|.*cProp$')), as.numeric)) %>%
    dplyr::select(-c(matches(rm_grps))) %>%
    filter(if_any(starts_with('SC'), ~!is.na(.))) %>%
    # filter(quality>75 & depth>40)  %>%
    mutate(filt_reason1 = case_when(quality<=75 | depth<=40 ~ 'Low quality',
                                    str_detect(chrom, mt_chrom_nm) ~ 'Mt'))
  
  print(dim(gt_fb_vtx_p1_bi))
  print(table(gt_fb_vtx_p1_bi$filt_reason1, useNA = 'ifany'))
  
  ##remove  mitochondrial calls
  # gt_fb_vtx_p1_bi <- gt_fb_vtx_p1_bi %>%
    # filter(!str_detect(chrom, 'Pf_M'))%>%
    # mutate(filt_reason1 = case_when(str_detect(chrom, 'Pf_M') ~ 'Mt'))
  
  print(dim(gt_fb_vtx_p1_bi))
  
  return(gt_fb_vtx_p1_bi)
}
```

```{r}
ss_afm_fb_vtx_p1bi <- pmap(list(unlist(gt_ss_afm_p1_bi, recursive = F), unlist(ss_afm_vtx_mtx, recursive = F)), ~fb_vt_bind(fb_gtype = ..1, vtx_mtx = ..2))
```


```{r}
## mark pass and fail snps

ss_po_pf7qc <- map(ss_afm_fb_vtx_p1bi, ~.x %>%
                            mutate(across(filt_reason1, ~case_when(is.na(.) ~ 'Pass',
                                           T ~ as.character(.)))
                                           )
)
```

### Visualize ratio of high VS low quality (based on freebayes) of SNPs {.tabset} 

```{r, fig.height=4, fig.width=15}
##Function for cleaner plot of SNPs falling in the different filter categories for each stage +/ strain cluster

# broadqc_lvls <- c('Pass', 'Absent in Pf7', 'Mt', 'Low quality')
broadqc_lvls <- c('Pass',  'Mt', 'Low quality')
broadqc_cols <- met.brewer('Isfahan2')[c(3:4,1)] %>% set_names(broadqc_lvls)


cln_plt_snp_filt_categ_pf7qc <- function(annot_snp_qc, filt_c = 'filt_reason1', xlb = 'Strain', qc_cols = broadqc_cols,  qc_lvls = broadqc_lvls, filt_rm=''){
  
##By retaining only SNPs with absolute difference between alt and ref >10 we also factor in depth per allele per cluster
annot_snp_qc %>% 
  filter(!(!!rlang::sym(filt_c) %in% filt_rm)) %>%
  mutate(across(!!rlang::sym(filt_c), ~droplevels(factor(., levels = qc_lvls)))) %>%
  # dplyr::count(groups, !!rlang::sym(filt_c), .drop=F) %>%
    dplyr::count(!!rlang::sym(filt_c), .drop=F) %>%
  # group_by(groups) %>% mutate(perc = round((n/sum(n))*100, digits = 1)) %>%
    mutate(perc = round((n/sum(n))*100, digits = 1)) %>%
  ggplot(aes(x=!!rlang::sym(filt_c),fill=!!rlang::sym(filt_c), y = n/1000)) + 
  geom_col(position = position_dodge(width = 1)) + 
  geom_text(aes(label=paste0(n, "(", perc,")"), color = !!rlang::sym(filt_c)), vjust = 0.5, hjust = -0.1, position = position_dodge(width = 1), size = 6.5, angle = 90, show.legend = F, fontface= 'bold') + 
  scale_color_manual(values = qc_cols) +
  scale_fill_manual(values = qc_cols) +
  scale_y_continuous(expand = expansion(mult =c(0, 0.8))) + 
  # facet_wrap(.~groups, nrow = 1,scales = "free_x", drop = T) +
  labs(x = xlb, y = 'Number of SNPs (X 1e3)', fill = 'SNP category') +
  theme_classic()+
  theme(axis.text.y = element_text(size = 14),
        axis.title.x = element_blank(),
        strip.text.x.top = element_text(size = 15, face='bold'),
        axis.text.x = element_blank(),
        axis.title = element_text(size = 18, face = 'bold'),
        legend.text = element_text(size = 16),
        legend.title = element_text(size = 18, face = 'bold'),
        legend.position="bottom",
        panel.border=element_rect(colour="grey",fill='transparent'),#,
        # plot.margin = unit(c(1,2,1,1),"cm")
        # legend.margin=margin()
      ) +
    # coord_fixed()+
    # coord_flip() +
    guides(fill = guide_legend( nrow = 1, title.position = "top"))
}
```


```{r}
ss_afm_broadqc_plt <- map(ss_po_pf7qc, ~cln_plt_snp_filt_categ_pf7qc(annot_snp_qc = .x, filt_c = 'filt_reason1', xlb = 'Strain', qc_cols = broadqc_cols,  qc_lvls = broadqc_lvls, filt_rm=''))

# ss_afm_broadqc_plt

```

```{r, results='asis'}

cat('\n')
  ss_afm_broadqc_plt %>% 
      purrr::iwalk(.,~{
        cat('#### ',.y,'   \n')
        print(.x)
        cat('\n \n')
      })
```


```{r}
##!!!! NOTE :Adding S to [G|D] [G|D|S] so as to detect the SC labeles in the k6 raw
##!!!! NOTE :Using instead of gt_fb_vtx_p1_bi_qc_pf7 since we want to only look at SNPs in pf 7 Mali
##By retaining only SNPs with absolute difference between alt and ref >10 we also factor in depth per allele per cluster
# snp_qc_tbl <- gt_fb_vtx_p1_bi_qc_pf7 %>% rename_with( ~str_replace(., '$','_raw'), matches("^[G|D|S].*_[A-Z]$")) %>% dplyr::select(chrom, pos,

snps_qc_filts <- function(m14_gt){
  
  snp_qc_tbl <- m14_gt %>% filter(filt_reason1 == 'Pass') %>% rename_with( ~str_replace(., '$','_raw'), matches("^(G|D|SC|P_|N).*(_A|_F|_M|_G|_Lo|_G|blet|tive|Unassigned|fail|[0-9])$")) %>% dplyr::select(chrom, pos, matches('.*_raw$|.*_qc$|.*_sNum$|.*_cProp$|.*_DP$|.*_RefN$|.*_AltN$'), alleles)   %>% 
      pivot_longer(.,cols = matches('^(G|D|SC|P_|N).*'), 
                   names_pattern = "(^[G|D|SC|P_|N].*[A|M|F|G|Lo|t|tive|Unassigned|stgQC_fail|(0-9)])_(alt_sNum|alt_cProp|ref_sNum|ref_cProp|raw|qc|DP|RefN|AltN)", 
                   names_to = c( "groups" ,".value")) %>%
      mutate(filt_reason = case_when(#filter != 'PASS' | is.na(filter) ~ 'Absent in Pf7', ##is.na(type) ~ 'Absent in pf7',### !!NB - Marking those missing in the pf7 dataset first. May need to be put last below if we want to include those absent if pf7 as some of these have been seen to be of good quality
                                     alt_sNum < 8 & ref_sNum < 8 ~  "Low UMI support",
                                     alt_sNum >= 15 & ref_sNum >= 15 ~  "Non distinct",
                                     abs(alt_sNum - ref_sNum) < 8 ~ 'Noisy' ,
                                     raw == '1' & alt_sNum > ref_sNum ~ 'Ref inverted' ,
                                     raw == '0' & ref_sNum > alt_sNum ~ 'Alt inverted',
                                     is.na(raw)&alt_sNum!=0&ref_sNum!=0  ~ "No fbayes call",
                                     # str_detect(filter, 'Hyper') ~ 'Hypervariable',
                                     # is.na(type) ~ 'Absent in pf7',
                                     T ~ 'Clean'
                                     # !is.na(qc) ~ "Clean"
                                     )
             ) %>%
    mutate(qc2 = case_when(filt_reason == 'Clean' ~ raw))
  
  return(snp_qc_tbl)

  }
```

```{r}
ss_po_qc <- map(ss_po_pf7qc, ~snps_qc_filts(m14_gt=.x))
# ss_po_qc[[1]] %>% str()
```


```{r, fig.height=5, fig.width=14}
# saveRDS(ss_afm_pf7M_qc, paste0("data/processed/",species, "/ss_afm_pf7M_qc.RDS"))
```



```{r, eval=FALSE}
## Make snp_qc directory for storing qc files
imap(ss_afm_pf7M_qc, ~system(paste0('mkdir data/processed/',species,'/', str_remove(.y, '\\..*'),'/', gtyp_step, '/',str_remove(.y, 'M.*\\.'),'_k', optimum_k[str_remove(.y, '\\..*')],'_', algn_nm,'/p1/snp_qc')))

imap(ss_afm_pf7M_qc, ~write_csv(.x, paste0('data/processed/',species,'/', str_remove(.y, '\\..*'),'/', gtyp_step, '/',str_remove(.y, 'M.*\\.'),'_k', optimum_k[str_remove(.y, '\\..*')],'_', algn_nm,'/p1/snp_qc/ss_afm_pf7M_qc.csv')))


```

### Visualize different filter categories of PASS SNPs (based on allele counts) {.tabset} 

```{r, fig.height=4, fig.width=20}
##Function to plot snps per the number of SNPs falling in the different filter categories for each stage + strain cluster
plt_snp_filt_categ <- function(annot_snp_qc, xlb = 'Strain'){
  
  annot_snp_qc %>% dplyr::count(groups, filt_reason) %>% 
  ggplot(aes(x=groups,fill=filt_reason, y = n)) + 
  geom_col(position = position_dodge(width = 1)) + 
  geom_text(aes(label=n, color = filt_reason), vjust = 0.5, hjust = -0.1, position = position_dodge(width = 1), size = 5, label.padding = unit(0.1, "lines"), angle = 90, show.legend = F) + 
  scale_y_continuous(expand = expansion(mult =c(0, 0.25))) + 
  labs(x = xlb, y = 'Number of SNPs', fill = 'SNP category') +
  theme_classic()+
  theme(axis.text.x = element_text(size = 16, angle = 90),
        axis.title = element_text(size = 18, face = 'bold'),
        legend.text = element_text(size = 16),
        legend.title = element_text(size = 18, face = 'bold')) 
  
}


```


```{r, fig.height=4, fig.width=20, results='hide'}
##Function to plot snps per the number of SNPs falling in the different filter categories for each stage + strain cluster
map(ss_po_qc, ~plt_snp_filt_categ(annot_snp_qc = .x))

```



```{r, fig.height=4, fig.width=15}
##Function for cleaner plot of SNPs falling in the different filter categories for each stage +/ strain cluster

snp_qc_lvls <- c('Clean','Low UMI support', 'Non distinct', 'Noisy', 'Ref inverted', 'Alt inverted')
snp_qc_cols <- c(met.brewer('NewKingdom')[c(2,3,1,4,5)],met.brewer('Navajo')[c(2)])  %>% set_names(snp_qc_lvls)


cln_plt_snp_filt_categ <- function(annot_snp_qc, xlb = 'Strain', ttl){
  
##By retaining only SNPs with absolute difference between alt and ref >10 we also factor in depth per allele per cluster
annot_snp_qc %>% 
  # filter(!filt_reason %in% c('Absent in Pf7', 'Low UMI support')) %>% 
  mutate(across(filt_reason, ~droplevels(factor(., levels = snp_qc_lvls)))) %>%
  dplyr::count(groups, filt_reason, .drop=F) %>%
    mutate(n_imp = as.numeric(case_when(filt_reason == 'Low UMI support' ~ NA, T ~ n)))%>%
  group_by(groups) %>% mutate(perc = round((n_imp/sum(n_imp, na.rm = T))*100, digits = 1)) %>%
  ggplot(aes(x=filt_reason,fill=filt_reason, y = n/1000)) + 
  geom_col(position = position_dodge(width = 1)) + 
  geom_text(aes(label=ifelse(is.na(n_imp), "", paste0(n_imp, "(", perc,")")), color = filt_reason), vjust = 0.5, hjust = -0.1, position = position_dodge(width = 1), size = 6.5, angle = 90, show.legend = F, fontface= 'bold') + 
  scale_color_manual(values = snp_qc_cols) +
  scale_fill_manual(values = snp_qc_cols) +
  scale_y_continuous(expand = expansion(mult =c(0, 0.05))) + 
  facet_wrap(.~groups, nrow = 1,scales = "free_x", drop = T) +
  labs(x = xlb, y = 'Number of SNPs (X 1e3)', fill = 'SNP category', title = ttl) +
  theme_classic()+
  theme(axis.text.y = element_text(size = 14),
        axis.title.x = element_blank(),
        strip.text.x.top = element_text(size = 15, face='bold'),
        axis.text.x = element_blank(),
        axis.title = element_text(size = 18, face = 'bold'),
        legend.text = element_text(size = 16),
        legend.title = element_text(size = 18, face = 'bold'),
        legend.position="bottom",
        panel.border=element_rect(colour="grey",fill='transparent'),
        plot.title = element_text(size = 20)#,
        # plot.margin = unit(c(1,2,1,1),"cm")
        # legend.margin=margin()
      ) +
    # coord_fixed()+
    # coord_flip() +
    guides(fill = guide_legend( nrow = 1, title.position = "top"))
}
```

```{r, fig.height=6, fig.width=30}
##By retaining only SNPs with absolute difference between alt and ref >10 we also factor in depth per allele per cluster
##!!!CHEATING - SORT OUT
ss_afm_k6r_qc_plt <- imap(ss_po_qc, ~cln_plt_snp_filt_categ(annot_snp_qc = .x, ttl = .y))
# ss_afm_k6r_qc_plt

```

```{r, results='asis'}

cat('\n')
ss_afm_k6r_qc_plt %>% 
      purrr::iwalk(.,~{
        cat('#### ',.y,'   \n')
        print(.x)
        cat('\n \n')
      })
```


```{r, fig.height=5, fig.width=14}
##Visualize and count the different classes of SNPs removed in QC
snp_filt_bias_viz <- function(annot_snp_qc, filt_cat){
  
  snps_filt_bias <- annot_snp_qc %>% dplyr::distinct(chrom, pos, filt_reason, groups)  %>% 
    # filter(filt_reason != 'Absent in Pf7') %>% 
    group_by(chrom,pos) %>% filter(!all(filt_reason == 'Low UMI support')) 
  
  snps_filt_bias_fc <- snps_filt_bias %>% filter(any(filt_reason == filt_cat)) %>% group_by(chrom, pos) %>% ungroup() 
  # snps_filt_bias_fc %>%  pivot_wider(names_from = groups, values_from = filt_reason) %>% View()
  snps_filt_bias_fc %>%  pivot_wider(names_from = groups, values_from = filt_reason) %>% dim() %>% print()
  snps_filt_bias_fc %>% filter(filt_reason == filt_cat) %>% dplyr::count(groups) %>% print()
  snps_filt_bias_fc %>%  dplyr::count(filt_reason, groups) %>% print()
  snps_filt_bias %>% filter(any(filt_reason == filt_cat)) %>% filter(any(filt_reason == 'Clean'))  %>% group_by(chrom, pos) %>% filter(sum(filt_reason == filt_cat) > 2) %>% ungroup() %>%  pivot_wider(names_from = groups, values_from = filt_reason) %>% dim %>% print()

}
```


```{r, fig.height=5, fig.width=14, results='hide'}
##Visualize and count the different classes of SNPs removed in QC
for (fc in c('Non distinct', 'Noisy', 'Alt inverted', 'Ref inverted')){
  map(ss_po_qc, ~snp_filt_bias_viz(annot_snp_qc = .x, filt_cat = fc))
}
```



```{r}
##Viz and Remove doublets and get informative SNPs
snp_rm_dblt_infomtv <- function(annot_snp_qc, clst2rmv = "Doublet"){
  annot_snp_qc %>% filter(str_detect(groups, clst2rmv)) %>% distinct(., pick(chrom,pos), .keep_all = T) %>% mutate(id =paste0(chrom,'_',pos)) %>% pull(id) %>% length() %>% print()

  annot_snp_qc_nd <- annot_snp_qc %>% filter(!str_detect(groups, clst2rmv))%>% filter(!is.na(qc2))
  
  annot_snp_qc_nd %>% group_by(chrom, pos) %>% filter(n_distinct(qc2) == 1 ) %>% distinct(., pick(chrom,pos), .keep_all = T) %>% dim() %>% print()
  
  annot_snp_qc_infomtv <- annot_snp_qc_nd  %>% group_by(chrom, pos) %>% filter(n_distinct(qc2) ==2 ) %>% ungroup
  
  annot_snp_qc_infomtv %>% distinct(., pick(chrom,pos), .keep_all = T) %>% dim() %>% print()


return(list(annot_snp_qc_nd, annot_snp_qc_infomtv))

}
```

```{r, fig.height=5, fig.width=14}
##By retaining only SNPs with absolute difference between alt and ref >10 we also factor in depth per allele per cluster

ss_afm_pf7M_qc_infomtv <- map(ss_po_qc, ~snp_rm_dblt_infomtv(annot_snp_qc = .x, clst2rmv = "None"))
```


```{r}
## Get genome size like this cut -f1,2 fasta.fai > genome.size
pf_chr_size = read_delim(chrm_sz, col_names = c('chrom', 'size')) %>% mutate(chr = str_remove_all(chrom,chrm_prfx)) 
```

IBS computation as advised by Mi

```{r}

gtypes_chr_cln<- function(snp_qc_fnl ){
  gtypes_chr_pos_pos <- snp_qc_fnl %>% dplyr::select(chrom, pos, groups, qc2, alleles) %>% pivot_wider(names_from = groups, values_from = qc2) %>% unite(., col="id", c(chrom, pos), sep = "_", remove = F) %>% mutate(chr = str_remove_all(chrom, chrm_prfx))
}
```



```{r}

ss_afm_pf7M_cln_gt<- map(ss_afm_pf7M_qc_infomtv, ~gtypes_chr_cln(snp_qc_fnl = .x[[1]]))
```


```{r}

saveRDS(ss_afm_pf7M_cln_gt, paste0("data/processed/",species, "/ss_afm_pf7M_cln_gt.RDS"))
```

################### END - CLEANING AND PREPROCESSING GENOTYPES #################

## VISUALIZATIONS & STRAIN VALIDATION
################### VISUALIZATIONS & STRAIN VALIDATION #########################

```{r, fig.height=3.8, fig.width=4.2}
##Heatmap function

gtypes_chr_cln_ibs_mtx <- function(gtypes_chr_pos){
  
  # ##Remove SC8 and SC1_F
  # gtypes_chr_pos <- snp_qc_tbl %>% filter(!(groups %in% c("SC8_F", "SC1_F")))%>% dplyr::select(chrom, pos, groups, qc2, alleles) %>% pivot_wider(names_from = groups, values_from = qc2) %>% unite(., col="id", c(chrom, pos), sep = "_", remove = F) %>% mutate(chr = str_remove_all(chrom, 'Pf3D7_0|Pf3D7_|_v3'))
  
  gtMat = gtypes_chr_pos %>% dplyr::select(-c(alleles,chrom, pos, chr))  %>% column_to_rownames("id") %>% mutate(across(everything(), ~as.numeric(str_replace(., '1', '2'))))%>% as.matrix() 
  gtMat = gtMat[,sort(colnames(gtMat))]
  
  
  gdsFile = tempfile()
    
  snpgdsCreateGeno(gdsFile, genmat = gtMat, sample.id = colnames(gtMat), 
                     snp.id = rownames(gtMat), snp.chromosome = gtypes_chr_pos$chr, 
                     snp.position = gtypes_chr_pos$pos, snp.allele = as.character(gtypes_chr_pos$alleles), snpfirstdim = TRUE)  
  
  gds = snpgdsOpen(gdsFile)
  
  gdsSub = snpgdsLDpruning(gds, ld.threshold = 0.2)
  
  ibs = snpgdsIBS(gds, num.thread = 4)
  
  rownames(ibs$ibs) = colnames(ibs$ibs) = colnames(gtMat)
    
  mat<-ibs$ibs
   
}
```

```{r, fig.height=3.8, fig.width=4.2}
##Heatmap function

gtypes_ibs_heatmap_one_don <- function(mtx, dendog = F){
# gtypes_ibs_heatmap <- function(mtx, dendo = F){
  
  colFun = suppressWarnings(brewer.pal(100, 'Greens'))
  colFun = circlize::colorRamp2(seq(0.5, 1, length.out = length(colFun)), 
                                  colFun)
   print(str(mtx))
   
   ##Remove columns and rows where everything in NA
   mtx <- mtx[,colSums(is.na(mtx)) < nrow(mtx)]
   mtx <- mtx[rowSums(is.na(mtx)) < ncol(mtx),]
  # fa = str_extract(colnames(mtx), 'MSC.*')
  # print(str(fa))
  
  # c_d = cluster_between_groups(mtx, fa)
  # dendog = if(dendo == T) c_d else F
  # dendog = if(dendo == T) cluster_between_groups(mtx, fa) else F
  # print(str(fa))
  # print(str(dendog))
  
  # fa = str_extract(colnames(mtx), 'MSC.*')
  # # fa_col = c("a" = 2, "b" = 3, "c" = 4)
  # dend1 = cluster_between_groups(mtx, fa)
  

  hm = Heatmap(mtx, name='IBS', col = colFun, show_row_names = TRUE, 
                 show_column_names = TRUE, show_row_dend = T, 
                 show_column_dend = T, row_title_rot = 0, cluster_rows = dendog, cluster_columns = dendog)
  
  
  hm2<-Heatmap(mtx, name='IBS', col = colFun, show_row_names = TRUE, 
                 show_column_names = TRUE, show_row_dend = T, 
                 show_column_dend = T, row_title_rot = 0, cluster_rows = dendog, cluster_columns = dendog, column_names_gp = gpar(fontsize = 15), row_names_gp = gpar(fontsize = 15), border='black',   heatmap_legend_param = list(title_gp = gpar(fontsize = 12, fontface = "bold"), labels_gp = gpar(fontsize = 12)),
                  layer_fun = function(j, i, x, y, width, height, fill) {
                      v = pindex(mtx, i, j)
                      l = v >0.98
                      grid.rect(x[l], y[l], width = width, height = height, gp = gpar(col = "#a4c2f4", lwd = 2, fill = "transparent"))
})
  
  print(hm)
  print(hm2)
  # print(hm3)
}
```

```{r, fig.height=4.1, fig.width=4.5}

ss_afm_ibs_mtx<- map(ss_afm_pf7M_cln_gt, ~gtypes_chr_cln_ibs_mtx(gtypes_chr_pos=.x))
ss_afm_ibs_hmap<- map(ss_afm_ibs_mtx, ~gtypes_ibs_heatmap_one_don(mtx=.x, dendo = F))
```

```{r, fig.height=5.1, fig.width=5.5}

ss_afm_ibs_hmap<- map(ss_afm_ibs_mtx, ~gtypes_ibs_heatmap_one_don(mtx=.x, dendo = T))
```

```{r}
##Function to creat pairwise comparison groups which are the used for plotting chromosome paints
strn_pw_comps <- function(gtypes_chr_pos, strn_regex){
  
  snp_pf7_afm <- gtypes_chr_pos %>% dplyr::select(id, chr, pos, starts_with(strn_regex)) #%>%  dplyr::select(!matches('^SC[0-9]_.$')) 
  
  col_pairs_strn <- snp_pf7_afm %>% dplyr::select(starts_with(strn_regex)) %>% colnames() %>% sort() %>% combn(., 2, simplify = F)
  
  strn_afm_comps <- map_dfc(col_pairs_strn, function(pair) {
    snp_pf7_afm %>%
      transmute(!!paste0(pair[1], "_vs_", pair[2]) := case_when(!is.na(.data[[pair[1]]]) & !is.na(.data[[pair[2]]]) & .data[[pair[1]]] == .data[[pair[2]]] ~ TRUE, 
                                                                !is.na(.data[[pair[1]]]) & !is.na(.data[[pair[2]]]) & .data[[pair[1]]] != .data[[pair[2]]] ~ FALSE))
  })
  
  strn_afm_comps <- base::cbind(snp_pf7_afm %>% dplyr::select(chr, pos), 
                                         strn_afm_comps)  %>%
    filter(!if_all(starts_with(strn_regex), ~is.na(.)))
  
  return(strn_afm_comps)
}
```


```{r}
##stage strain pairwise comparison groups
ss_afm_pf7M_afm_comps <- map(ss_afm_pf7M_cln_gt, ~strn_pw_comps(gtypes_chr_pos = .x, strn_regex = c('SC','N')))

```

```{r}
##Function to create binned windows across genome and compare pairwise strains and generate the plotting details for chromosome painting
b_strn_c_wndw_fn <- function(strn_c_tbl, wndw = 100000) {
  strn_c_tbl %>% 
    group_by(chr) %>% 
    filter(!is.na(chr)) %>%
    mutate(pos_range = cut(pos, breaks = c(-Inf,seq(0, max(pos), by = wndw),Inf))) %>% 
    pivot_longer(., cols = starts_with(c('SC','N')), names_to = 'comps', values_to = 'sim') %>% 
    dplyr::count(comps, chr, pos_range, sim) %>% 
    group_by(chr, comps, pos_range) %>% 
    mutate(overlap_TF = ifelse(any(is.na(sim)), sum(!is.na(sim)), n_distinct(sim)))%>%
    ungroup() %>% 
    filter(!(is.na(sim) & overlap_TF>0))  %>% 
    mutate(rltd = case_when(overlap_TF == 0 ~ 'No SNP', overlap_TF == 2 ~ 'Both', overlap_TF==1& sim==TRUE ~ 'Same', overlap_TF==1&sim==FALSE ~ 'Different')) %>% 
    group_by(chr, comps, pos_range, rltd) %>% 
    arrange(dplyr::desc(sim), .by_group=T) %>%
    summarise(overlap_snps_No = paste(n, collapse = ","), 
              overlap_snps_prop = log10(n[1]/n[2]))  %>% 
    mutate(across(overlap_snps_No, ~ifelse(rltd == 'No SNP', NA, .))) %>%
    ungroup()  %>%
    mutate(normalized_snps_prop = case_when(rltd == "Same" ~ max(overlap_snps_prop, na.rm = TRUE), 
                                            rltd == "Different" ~ min(overlap_snps_prop, na.rm = TRUE), 
                                            # rltd == 'No SNP' ~ NA, 
                                            T ~ overlap_snps_prop)) %>%
    mutate(start = as.numeric(str_extract(pos_range, "\\d+.*(?=,)")), end = as.numeric(str_extract(pos_range, "(?<=,)\\d+.*(?=])"))) %>% 
    left_join(.,pf_chr_size, by =c('chr')) %>%
    mutate(across(chr, ~factor(., levels = c(1:14)))) 
}
```


```{r}
##stage strain pairwise comparison chromosome binning of SNPs and creation of windows for plotting
ss_afm_wndws <- map(ss_afm_pf7M_afm_comps, ~b_strn_c_wndw_fn(strn_c_tbl = .x))

```



```{r}
## Function for chromosome painting

p_chart_shared_snps <- function(gtypes_comps_tbl = b_strn, ttl, lg = "right") {

      
    gtypes_comps_tbl %>% 
      separate(.,overlap_snps_No, into=(c('same2', 'diff2')), sep=',' ) %>% 
      pivot_longer(cols = c('same2','diff2'), names_to = 'shared2', values_to = 'shared_n') %>% 
      filter(!is.na(shared_n)) %>%
      mutate('Shared' = case_when(rltd == 'Both' & shared2 == 'same2' ~ 'Same', 
                                  rltd == 'Both' & shared2 == 'diff2' ~ 'Different', 
                                  rltd == 'Different' & shared2 == 'same2' ~ 'Different', 
                                  T ~ rltd)) %>%
      mutate(across(shared_n, as.numeric)) %>%
      group_by(comps, Shared) %>% 
      summarise('snpsNo' = sum(shared_n))%>%
      mutate(snpsProp = snpsNo/sum(snpsNo)) %>%
      arrange(dplyr::desc(snpsProp), .by_group = T) %>%
  separate(comps, into=(c('pb_grp1', 'pb_grp2')), sep='_vs_' ) %>% 
    ggplot(., aes(x="", y=snpsProp, fill=Shared))+
    geom_bar(width = 1, stat = "identity") +
    geom_text(aes(y = ifelse(Shared == 'Same', 0.2, 0.8),
                  label = snpsNo), size=3.5) +
  facet_grid(pb_grp1~pb_grp2) +
    labs(title = ttl)+
    theme_classic() +
      theme(legend.position = lg)


}


```

Pie of snps compared
```{r, warning=FALSE}
# strain stage (afm) chromosome painting
pp_plts <- imap(ss_afm_wndws, possibly(~p_chart_shared_snps(gtypes_comps_tbl =.x, ttl=.y, lg = "none"), 'No overlapping SNPs between the 2 groups'))

# ss_afm_chromp_smth[1:3]
```

```{r, warning=FALSE, fig.width=8, fig.height=8}
# strain stage (afm) chromosome painting
pp_plts
```

Notes on Confirmed K

MSC27	1, MSC35	3, MSC42 1, MSC44	3, MSC47	6, MSC36	2, MSC16	2, MSC15	2 (The numbers are low - could be k=3 - i only investigated clusters with >20 cells but this can be reduced to >10)

Manually assign negative for which we assessed to the relevant Ks
MSC27 = k4 : Neg==SC2
MSC49 = k6 : Neg==SC1
MSC60 = k7 : Neg==SC3
MSC66 = k8 : Neg==SC5
MSC68 = k4 : Neg==SC2


##Simpler chromosome plots(no continous gradient) for SS_AG and S

```{r}
## Function for chromosome painting

chrom_paintr <- function(gtypes_comps_tbl = b_strn_c, ttl, expn=100000, x_font_sz = 13) {

      gs <-gtypes_comps_tbl %>% mutate(across(start, ~(./expn)),across(end, ~(./expn)), across(size, ~(./expn))) %>%
      ggplot()+
      geom_segment(
    aes(x = 0-(2000/expn), xend = size+(2000/expn), y = chr, yend = chr),
    linewidth = 3.5, lineend = "round", colour = "black"
  ) + 
      geom_segment(
    aes(x = 0, xend = size, y = chr, yend = chr),
    linewidth = 2.8, lineend = "round", colour = "white"
  ) +
  geom_segment(
    aes(
      x = start, xend = end,
      y =  chr, yend =  chr,
      colour = rltd
    ),
    linewidth = 2.8
  ) +
  # scale_color_manual(values = c('Same'= '#D4AF37', 'Different'= '#2C3E50','Both'='lightgrey', 'No SNP'='white'))+      
  scale_color_manual(values = c('Same'= '#DACB7D', 'Different'= '#2C3E50','Both'='#A9B7C5', 'No SNP'='white'))+ ##prev colours "black" ,"#ffd94d" 
  # facet_wrap(comps~., ncol = 5) +
      labs(title = ttl) +
      labs(colour = "Allele \ncomparison", x = 'Position (bases)', y ="Chromosome") +
  theme_classic() +
  theme(axis.text.y = element_text(size =  13),
        axis.text.x = element_text(size =  x_font_sz),
        axis.title = element_blank(),
        title = element_text(size = 12, face = "bold"),
        legend.position = "none") 
    

  }

```


strains split by stage Without colour gradients
```{r, warning=F}
# strain stage (afm) chromosome painting
ss_afm_chromp <- map(ss_afm_wndws, ~{
  smp_tbl <- .x
  smp_tbl %>% split(., ~factor(comps)) %>% imap(.,possibly(~chrom_paintr(gtypes_comps_tbl =.x, ttl=.y), 'No overlapping SNPs between the 2 groups'))
})

```
CONFIRMATION OF STRAINS
33=8, 
```{r, warning=F}
# strain stage (afm) chromosome painting
names(ss_afm_chromp)

```

```{r, warning=F}
# strain stage (afm) chromosome painting
ss_afm_chromp[str_detect(names(ss_afm_chromp), "MSC47.strain.k6")]

```


```{r, eval=T}
## Function for chromosome painting

chrom_paintr_smth_gradient <- function(gtypes_comps_tbl = b_strn, ttl, expn=100000, x_font_sz = 13, lg = "right") {

      # gs <-gtypes_comps_tbl %>% mutate(across(start, ~(./expn)),across(end, ~(./expn)), across(size, ~(./expn))) %>%
      gs <-gtypes_comps_tbl %>% mutate(across(start, ~(./expn)),across(end, ~(./expn)), across(size, ~(./expn))) %>%
      ggplot()+
        geom_segment(
    aes(x = 0-2000/expn, xend = size+2000/expn, y = chr, yend = chr),
    linewidth = 3.5, lineend = "round", colour = "black"
  ) + 
      geom_segment(
    aes(x = 0, xend = size, y = chr, yend = chr),
    linewidth = 2.5, lineend = "round", colour = "white"
  ) +
  geom_segment(
    aes(
      x = start, xend = end,
      y =  chr, yend =  chr,
      colour = normalized_snps_prop#,
      # colour = rltd
    ),
    linewidth = 2.5
  ) +
  # scale_color_manual(values = c('Same'='black', 'Different'='#ffd94d', 'Both'='lightgrey', 'No SNP'='white'))+
  scale_color_gradient2(low="black" , mid = "#F5F5F5", high ="#ffd94d" ,
                        limits = c(min(gtypes_comps_tbl$normalized_snps_prop, na.rm = TRUE), max(gtypes_comps_tbl$normalized_snps_prop, na.rm = TRUE)),
                        breaks = seq(min(gtypes_comps_tbl$normalized_snps_prop, na.rm = TRUE), max(gtypes_comps_tbl$normalized_snps_prop, na.rm = TRUE), length.out = 30),
                        oob = scales::squish,
                        na.value = "white",
                        midpoint = 0,
                        guide = lg) +
  # geom_label_repel(aes(label=new_column, x=(start + end)/2, y = comps), size=3,label.padding = unit(0.01, "lines"),alpha = 1,label.size = 0.01, color = alpha('black', .2))+
  # facet_wrap(comps~., ncol = 5) ++
      labs(title = ttl) +
  theme_classic() +
  theme(strip.text = element_text(size = 12, face = 'bold'),
        axis.text.y = element_text(size = 12),
        axis.text.x = element_text(size =  x_font_sz)) 
    
    pp <- gtypes_comps_tbl %>% 
      separate(.,overlap_snps_No, into=(c('same2', 'diff2')), sep=',' ) %>% 
      pivot_longer(cols = c('same2','diff2'), names_to = 'shared2', values_to = 'shared_n') %>% 
      filter(!is.na(shared_n)) %>%
      mutate('Shared' = case_when(rltd == 'Both' & shared2 == 'same2' ~ 'Same', 
                                  rltd == 'Both' & shared2 == 'diff2' ~ 'Different', 
                                  rltd == 'Different' & shared2 == 'same2' ~ 'Different', 
                                  T ~ rltd)) %>%
      mutate(across(shared_n, as.numeric)) %>%
      group_by(Shared) %>% 
      summarise('snpsNo' = sum(shared_n))%>%
      arrange(dplyr::desc(snpsNo), .by_group = T) %>%
    ggplot(., aes(x="", y=snpsNo, fill=Shared))+
    geom_bar(width = 1, stat = "identity") + 
    coord_polar("y", start=0)+
    geom_text(aes(y = snpsNo/2 + c(0, cumsum(snpsNo)[-length(snpsNo)]), 
                  label = snpsNo), size=5) +
    theme_void() +
      theme(legend.position = lg)

    ggdraw(gs + theme_half_open(12)) +
      draw_plot(pp, .5, .11, .36, .36) 
}


```


strains split by stage Without colour gradients
```{r, warning=FALSE, eval=T}
# strain stage (afm) chromosome painting
ss_afm_chromp_smth <- map(ss_afm_wndws, ~{
  smp_tbl <- .x
  smp_tbl %>% split(., ~factor(comps)) %>% imap(.,possibly(~chrom_paintr_smth_gradient(gtypes_comps_tbl =.x, ttl=.y, lg = "none"), 'No overlapping SNPs between the 2 groups'))
})

# ss_afm_chromp_smth[1:3]
```


```{r, warning=FALSE, eval=T}
ss_afm_chromp_smth["MSC47.strain.k6"]
```

```{r, warning=FALSE, eval=T}
ss_afm_chromp_smth["MSC35.strain.k3"]
```


```{r, warning=FALSE, eval=T}
## Check those thought to be k=1
ss_afm_chromp_smth["MSC27.strain.k2"]
```
```{r, warning=FALSE, eval=T}
## Check those thought to be k=1
ss_afm_chromp_smth["MSC42.strain.k2"]
```
```{r, warning=FALSE, eval=T}
## Check those thought to be k=1
ss_afm_chromp_smth["MSC35.strain.k3"]
```

```{r, warning=FALSE, eval=T}
ss_afm_chromp_smth["MSC44.strain.k3"]
```


Notes on Confirmed K - Manually assign negative for which we assessed to the relevant Ks

MSC27	2 (Neg == SC1) :not very clear going with 2 for now.
MSC35	3 (Neg == SC2) : going with 3 for now. very little differences between strains. could be inbreeding or highly variable genes (vars etc)
MSC42 1 
MSC44	3 (Neg == SC1) : 
MSC47	6 (Neg == SC1) : 
MSC36	2
MSC16	2
MSC15	3

```{r, warning=FALSE, eval=F}
## Write out the strain assignments
## Get optimum strain K based on assesment above
optimum_k_qcd <- c("MSC27" = 2, "MSC35" = 3, "MSC42" = 1, "MSC44" = 3, "MSC47" = 6, "MSC36" = 2, "MSC16" = 2, "MSC15" = 3) %>% .[sample_name_nms]
optimum_k_qcd <- optimum_k_qcd[sort(names(optimum_k_qcd))]

##updated strain metadata
strn_c_tbl <- readRDS(paste0("data/processed/",species, "/strn_c_tbl.RDS"))
strn_c_tbl_min <- strn_c_tbl[str_detect(names(strn_c_tbl), "_minmap")]
names(strn_c_tbl_min) <-  str_remove(names(strn_c_tbl_min), "_minmap")
```


```{r, warning=FALSE, eval=T}

for (i in names(strn_c_tbl_min)) {
        if (optimum_k_qcd[[i]] == 1) {
          strn_c_tbl_min[[i]][,"StrainQCd"] = "SC0"
        } else if (optimum_k_qcd[[i]] > 1) {
          strn_c_tbl_min[[i]][,"StrainQCd"] = strn_c_tbl_min[[i]][,paste0('Strain_hmo_', str_remove(i, 'SC'), "_minmap_",optimum_k_qcd[[i]])]
        } 
    
  }

```

```{r, warning=FALSE, eval=T}

## REvise negative strains and assign estimated strain based on plots above

strn_c_tbl_min[["MSC27"]][,"StrainQCd"] <- strn_c_tbl_min[["MSC27"]] %>% 
  mutate(StrainQCd = case_when(StrainQCd == "Negative" ~ "SC1",
                                   T ~ StrainQCd)) %>% select(StrainQCd)

strn_c_tbl_min[["MSC35"]][,"StrainQCd"] <- strn_c_tbl_min[["MSC35"]] %>% 
  mutate(StrainQCd = case_when(StrainQCd == "Negative" ~ "SC2",
                                   T ~ StrainQCd))%>% select(StrainQCd)

strn_c_tbl_min[["MSC44"]][,"StrainQCd"] <- strn_c_tbl_min[["MSC44"]] %>% 
  mutate(StrainQCd = case_when(StrainQCd == "Negative" ~ "SC1",
                                   T ~ StrainQCd))%>% select(StrainQCd)

strn_c_tbl_min[["MSC47"]][,"StrainQCd"] <- strn_c_tbl_min[["MSC47"]] %>% 
  mutate(StrainQCd = case_when(StrainQCd == "Negative" ~ "SC1",
                                   T ~ StrainQCd))%>% select(StrainQCd)


```

```{r, warning=FALSE, eval=T}
saveRDS(strn_c_tbl_min, paste0("data/processed/",species, "/strn_c_tbl_min_qcd.RDS"))
```

################### END CHROMOSOME PAINTING & STRAIN VALIDATION ################

## BETWEEN DONOR IBS
################### BETWEEN DONOR IBS###########################################


```{r, warning=FALSE, eval=T}
## Get donors with more than 1 strain
optimum_k_polyk <- optimum_k_qcd[optimum_k_qcd > 1]
```

```{r, fig.height=4.1, fig.width=4.5, eval=T}
## Get only the genotype tables for the optimum k 
## !!!NOTE - CHEATING - We need to genotype the strains that have K=1 so as to include them in the between donor genotype comparisons
ss_afm_pf7M_cln_gt_polyk <- ss_afm_pf7M_cln_gt[unlist(imap(optimum_k_polyk, ~paste0(.y, ".strain.k", .x)))]

```


```{r}
## !!!NOTE - CHEATING - Another round of pseudobulk genotyping after manually assigning Negatives to strains is necessary - here we just remove them
## !!!NOTE - For QCD genotypes this should not be necessary for negatives unless they are genuine negative that can't be collapsed with strain manually
## Remove doublets and negatives
# ss_afm_pf7M_qc <- map(ss_afm_pf7M_qc, ~.x %>% filter(!str_detect(groups, "Doublet|Negative")))
ss_afm_pf7M_cln_gt_polyk <- map(ss_afm_pf7M_cln_gt_polyk, ~.x %>% select(-c(contains("Doublet"), contains("Negative"))))

```

```{r, fig.height=4.1, fig.width=4.5, eval=T}
## All genotypes done separately since when done together the error for DUPLICATE colnames appears

ss_cln_gt_tbl_s <- ss_afm_pf7M_cln_gt_polyk[c(seq(1,length(ss_afm_pf7M_cln_gt_polyk), length(grps)))] %>%
    imap(~ rename_with(.x, function(x) paste(.y, x, sep = "_"), -c(id,chrom, pos, alleles, chr))) %>%
    purrr::reduce(left_join, by = c('id','chrom', 'pos', 'alleles', 'chr')) %>% 
  rename_with(~str_remove_all(., '\\.st.*ain|k[0-9]+_'), starts_with(c("MSC")) )%>% 
  rename_with(~str_replace_all(., c('MSC'='M', '\\.'='_')), starts_with(c("MSC")) )

ss_cln_gt_tbl_s <- ss_cln_gt_tbl_s[,sort(colnames(ss_cln_gt_tbl_s))]

## 

ss_cln_all_mtx <- list(ss_cln_gt_tbl_s) %>% set_names(c("all.strain_qcd"))

```


```{r, fig.height=3.8, fig.width=4.2}
##Heatmap function

gtypes_chr_cln_ibs_mtx <- function(gtypes_chr_pos){
  
  gtMat = gtypes_chr_pos %>% dplyr::select(-c(alleles,chrom, pos, chr))  %>% column_to_rownames("id") %>% mutate(across(everything(), ~as.numeric(str_replace(., '1', '2'))))%>% as.matrix() 
  gtMat = gtMat[,sort(colnames(gtMat))]
  
  
  gdsFile = tempfile()
    
  snpgdsCreateGeno(gdsFile, genmat = gtMat, sample.id = colnames(gtMat), 
                     snp.id = rownames(gtMat), snp.chromosome = gtypes_chr_pos$chr, 
                     snp.position = gtypes_chr_pos$pos, snp.allele = as.character(gtypes_chr_pos$alleles), snpfirstdim = TRUE)  
  
  gds = snpgdsOpen(gdsFile)
  
  gdsSub = snpgdsLDpruning(gds, ld.threshold = 0.2)
  
  ibs = snpgdsIBS(gds, num.thread = 4)
  
  rownames(ibs$ibs) = colnames(ibs$ibs) = colnames(gtMat)
    
  mat<-ibs$ibs
   
}
```

Multisample heatmap


```{r, fig.height=15, fig.width=20}

gt_all_ibs_mtx<- map(ss_cln_all_mtx, ~gtypes_chr_cln_ibs_mtx(gtypes_chr_pos=.x))

```


```{r}

## This will replace the above once the donor labels msc are capitalized
clustr_numbers_s <- map(strn_c_tbl_min, ~.x %>% 
          add_count(StrainQCd, name = "strain_n") %>% 
        distinct(StrainQCd, strain_n) %>% 
        unite(col = 'cluster', StrainQCd, sep = '_'))  %>% 
    bind_rows(., .id='sample') %>%
  mutate(across(sample, ~str_remove_all(., 'SC')))%>% 
    unite(col = 'sample_cluster', sample, cluster, sep = '_')#%>%
  #mutate(across(sample_cluster, ~str_replace(., "MSC", "M")))

# rownames(clustr_numbers_s) <- NULL
```


```{r}
## Read in msc annotations 
donor_cols = met.brewer(name = "Renoir")[1:length(sample_name_nms)] %>% set_names(str_remove(sample_name_nms, "SC"))
```




```{r, fig.height=8, fig.width=10.5}
##Merge strain numbers to IBD results
ibs_don_hm <- function(ibsmtx, clustr_numbers, strn_cat, start_no = 0.5, test_nm = "IBS"){
colFun = suppressWarnings(brewer.pal(100, 'Greens'))
colFun = circlize::colorRamp2(seq(start_no, 1, length.out = length(colFun)), 
                                  colFun)
  
sibd_pht_mtx <- ibsmtx
print(colnames(sibd_pht_mtx))
sibd_pht_mtx <- sibd_pht_mtx[,row.names(sibd_pht_mtx)]

# fa = str_extract(colnames(sibd_pht_mtx), 'ms.*[0-9]')
# fa = str_remove(colnames(sibd_pht_mtx), '_qcd.*')
fa = str_remove(colnames(sibd_pht_mtx), '_.*')
c_d = cluster_within_group(sibd_pht_mtx, fa)

print(fa)
# fr = str_extract(rownames(sibd_pht_mtx), 'ms.*[0-9]')
# fr = str_remove(rownames(sibd_pht_mtx), '_qcd.*')
fr = str_remove(rownames(sibd_pht_mtx), '_.*')
# print(fr)
c_r = cluster_within_group(sibd_pht_mtx, fr)

# nms = unique(str_extract(colnames(sibd_pht_mtx), 'ms.*[0-9]'))
nms = unique(str_remove(colnames(sibd_pht_mtx), '_qcd.*'))
# sample_c_cols <- met.brewer("Thomas")[1:length(nms)] %>% set_names(nms)
sample_c_cols <- donor_cols

# sibd_pht_mtx_cln <- sibd_pht_mtx %>% as.data.frame()%>% rownames_to_column('sid') %>% mutate(across(sid, ~str_replace_all(., c('G_MSC' = 'M')))) %>% rename_with(~str_replace_all(., c('G_MSC' = 'M')), starts_with('SC')) %>% column_to_rownames('sid') %>% as.matrix()
sibd_pht_mtx_cln <- sibd_pht_mtx %>% as.data.frame()%>% rownames_to_column('sid') %>% 
  #mutate(across(sid, ~str_replace_all(., c('MSC' = 'M', '_qcd'='')))) %>% rename_with(~str_replace_all(., c('MSC' = 'M', '_qcd'='')), starts_with('MSC')) %>% 
  left_join(., clustr_numbers, by = c('sid'='sample_cluster')) %>% 
  mutate(across(sid, ~paste0(., ' [', !!rlang::sym(strn_cat), ']'))) %>% select(-c(!!rlang::sym(strn_cat))) %>%
  column_to_rownames('sid') %>% 
  as.matrix()

print(head(sibd_pht_mtx_cln))
  
sibd_pht<- Heatmap(sibd_pht_mtx_cln, name=test_nm, 
                   col = colFun, 
                   show_row_names = TRUE, 
                   row_names_side = c("left"), 
                   show_column_names = TRUE,  
                   show_row_dend = T,  
                   show_column_dend = F, 
                   row_title_rot = 0, 
                   cluster_rows = c_r, 
                   clustering_method_rows = "complete",
                   cluster_columns = c_d, 
                   column_names_gp = gpar(fontsize = 15), 
                   row_names_gp = gpar(fontsize = 15), 
                   heatmap_legend_param = list(title_gp = gpar(fontsize = 15, fontface = "bold"), labels_gp = gpar(fontsize = 15)),
                   bottom_annotation = HeatmapAnnotation(Donor = fa, 
                                                         col = list(Donor = sample_c_cols), 
                                                         annotation_name_gp = gpar(fontsize = 15, fontface = 'bold'), 
                                                         annotation_legend_param = list(title_gp = gpar(fontsize = 15, fontface = "bold"), labels_gp = gpar(fontsize = 15))),
                   column_split = length(nms), 
                   border = TRUE, 
                   left_annotation=rowAnnotation(Donor = fa, col = list(Donor = sample_c_cols), annotation_name_gp = gpar(fontsize = 15, fontface = 'bold'), show_legend = F), row_split = length(nms), 
                   column_title = NULL,
                   column_title_gp = gpar(fontsize = 16), 
                   row_title = NULL,
                   row_title_gp = gpar(fontsize = 16)
              # border='black',row_gap=unit(.015, "npc"), column_gap=unit(.015, "npc"), 
              ) %>%
  draw(., merge_legend = TRUE)

sibd_pht
}
```



```{r, fig.height=18, fig.width=18}

hms<- pmap(list(gt_all_ibs_mtx, list( clustr_numbers_s), c("strain_n" )), ~ibs_don_hm(ibsmtx=..1, clustr_numbers=..2, strn_cat=..3))

```
################### END - BETWEEN DONOR IBS########################################

## Write out for genotyping

Write out strain_qcd + stage_ag for pseudobulk genotyping

```{r, warning=FALSE, eval=T}
## Get donors with more than 1 strain
optimum_k_polyk <- optimum_k_qcd[optimum_k_qcd > 1]
# optimum_k_polyk <- optimum_k_polyk[!names(optimum_k_polyk) %in% c("MSC37", "MSC38")]
```

```{r, eval=F}
## Write out the clean genotypes for final genotyping
sv_dir = "pbulk_gtypes_postQC_cln"

# for (s in names(strn_c_tbl_min)) {
for (s in names(optimum_k_polyk)) {
  # narrow_k <- opt_narrow_clusters_nms[[str_remove(s, "_[a-z]+")]]
  narrow_k <- optimum_k_polyk[s]
  ## Assign minimap/hsat mapper directory
  # algn_nm = str_extract(s, "hsat|minmap|lr_ref")
  algn_nm = "minmap"
  # s_nm = str_remove(s, "_[a-z]+")
  s_nm = s
  
  print(algn_nm)
  print(s_nm)
  
  ## !!!NOTE - Dangerous command - activate only when barcodes are modified
  # system(paste0('rm -r data/processed/',species,'/',s_nm,'/', sv_dir,'/bcodes/',algn_nm,'/st*'))
  
  for (k in narrow_k) {
    print(narrow_k)
    # for (str_q in c('StrainQCd')){
      
      ss_qcd_ano <- map(strn_c_tbl_min[s], ~.x %>% rename("str" = StrainQCd) %>% group_by(str) %>% filter(n() >= 20) %>% ungroup() %>% mutate(across(c(str), ~droplevels(factor(.)))) %>% split(f = ~ str, drop = T))
      
      names(ss_qcd_ano[[1]]) <- names(ss_qcd_ano[[1]]) %>% str_replace(., '\\.', '_')
      # print(str(ss_qcd_ano[[1]]))

      system(paste0('mkdir -p data/processed/',species,'/',s_nm,'/', sv_dir,'/bcodes/', algn_nm,'/strain_k',k,'/'))
      print(paste0('mkdir -p data/processed/',species,'/',s_nm,'/', sv_dir,'/bcodes/', algn_nm,'/strain_k',k,'/'))

      imap(ss_qcd_ano[[1]], ~write(.x$bcode, paste0('data/processed/',species,'/',s_nm,'/', sv_dir,'/bcodes/',algn_nm,'/strain_k',k,'/',.y, '.tsv')))

      # for(stg_c in c('stage_afm', 'stage_ag')){
      # for(stg_c in c('stage_ag')){
        
        ss_qcd_ano <- map(strn_c_tbl_min[s], ~.x %>% rename("str" = StrainQCd, "stg" = stage_ag) %>% group_by(str, stg) %>% filter(n() >= 20) %>% ungroup() %>% mutate(across(c(str, stg), ~droplevels(factor(.)))) %>% split(f = ~ str+stg, drop = T))
        names(ss_qcd_ano[[1]]) <- names(ss_qcd_ano[[1]]) %>% str_replace(., '\\.', '_')
        print(names(ss_qcd_ano[[1]]))

        system(paste0('mkdir -p data/processed/',species,'/',s_nm,'/', sv_dir,'/bcodes/',algn_nm,'/stage_ag_strain_k',k,'/'))
        print(paste0('mkdir -p data/processed/',species,'/',s_nm,'/', sv_dir,'/bcodes/',algn_nm,'/stage_ag_strain_k',k,'/'))

        imap(ss_qcd_ano[[1]], ~write(.x$bcode, paste0('data/processed/',species,'/',s_nm,'/', sv_dir,'/bcodes/',algn_nm,'/stage_ag_strain_k',k,'/',.y, '.tsv')))

    #   }
    # }
  
}

}
```

```{r, eval=FALSE}
## Remove empty slots where there are no overlapping SNPs between the 2 groups being compared
pts <- ss_afm_chromp_smth[[1]][unlist(map(ss_afm_chromp_smth[[1]], ~length(class(.x)) == 2))]

chrom_pdf_save <- function(plt_list = pts, No_plots_p_page =24, No_cols = 4, don = "MSC33", grp = "stage_afm_strain") {
  print(length(plt_list))
  plt_list <- plt_list[unlist(map(plt_list, ~length(class(.x)) == 2))]
  print(length(plt_list))
  
  plot_lists_sep <- split(plt_list, (seq_along(plt_list) - 1) %/% No_plots_p_page)
  
  print(don)
  
  system(paste0("mkdir -p plots/pbulk_gtypes_preQC/", don))
  
  pdf(paste0("plots/pbulk_gtypes_preQC/", don, "/", grp, ".pdf"),  onefile = TRUE, width=12, height = 16.91567) ## A4 aspect ratio (1.409639) but bigger
  for (i in seq_along(plot_lists_sep)) {
    print(i)
    grid.arrange(grobs = plot_lists_sep[[i]], ncol = No_cols)
  }
  dev.off()
}
```


```{r, eval=FALSE}
## Remove empty slots where there are no overlapping SNPs between the 2 groups being compared
imap(ss_afm_chromp_smth, ~chrom_pdf_save(plt_list = .x, don = str_remove(.y, "\\..*"), grp = paste0(str_replace(.y, ".", "_"),"k")))
```
